<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HandyCNV: An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH  • HandyCNV</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="HandyCNV: An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH ">
<meta property="og:description" content="The package is aimed to provide a Standard, Reproducible and Time-saving pipelines for the post-analysis of CNV, CNVR and ROH.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">HandyCNV</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/HandyCNV.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="welcome-to-handycnv" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#welcome-to-handycnv" class="anchor"></a>Welcome to HandyCNV</h1></div>

<p>An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH</p>
<p align="center">
<img src="https://github.com/JH-Zhou/HandyCNV/blob/master/man/figures/High_resolution_pipeline_HandyCNV.jpg"></p>
<p align="center">
Main functions and outputs form HandyCNV
</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package was originally designed for the Post-analysis of CNV results inferred from PennCNV and CNVPartition (GenomeStudio). However, it has now been expanded to accept input files in standard formats for a wider range of applications. Our motivation is to provide a standard, reproducible and time-saving pipeline for the post-analysis of CNVs and ROHs detected from SNP genotyping data for the majority of diploid Species. The functions provided in this package can be categorised into five sections: Conversion, Summary, Annotation, Comparison and Visualization. The most useful features provided are: integrating summarized results, generating lists of CNVR, annotating the results with known gene positions, plotting CNVR distribution maps, and producing customised visualisations of CNVs and ROHs with gene and other related information on one plot. This package also supports a range of customisations, including the colour, size of high resolution figures, and output folder, avoiding conflict between the results of different runs. Running through all functions detailed in the vignette could help us to identify and explore the most interesting genomic regions more easily.</p>
</div>
<div id="vignettes-and-manual" class="section level1">
<h1 class="hasAnchor">
<a href="#vignettes-and-manual" class="anchor"></a>Vignettes and Manual</h1>
<p>The details examples please visit our Github pages: <a href="https://jh-zhou.github.io/HandyCNV/" class="uri">https://jh-zhou.github.io/HandyCNV/</a></p>
</div>
<div id="installation-and-prerequisites" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-prerequisites" class="anchor"></a>Installation and Prerequisites</h1>
<p>First, to run this package, we need to make sure that R (Version &gt;= 3.5.2) is installed in your computer (R download link: <a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a>). Once R is installed, the ‘HandyCNV’ package can be installed from Github repository by running the following script. If you rarely used R, it may take more time to install the ‘HandyCNV’ for the first time.</p>
<div id="1-method-one-install-from-github-repo-directly" class="section level2">
<h2 class="hasAnchor">
<a href="#1-method-one-install-from-github-repo-directly" class="anchor"></a>1. Method one, install from Github Repo directly</h2>
<pre class="{r}"><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("remotes") # Run this code if you haven't install 'remotes' package before 
remotes::install_github(repo = "JH-Zhou/HandyCNV@v1.1.7")</a></code></pre>
</div>
<div id="2-method-two-install-manually" class="section level2">
<h2 class="hasAnchor">
<a href="#2-method-two-install-manually" class="anchor"></a>2. Method two, install manually</h2>
<p>If the first method cannnot work well for some reasons, we can manually download the ‘Source code (Zip)’ from the newly released tag at here: <a href="https://github.com/JH-Zhou/HandyCNV/releases/tag/v1.1.7">Download Source Code</a></p>
<p>Then install the Source Code from the local path by following code:</p>
<pre class="{r}"><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("remotes") # Run this code if you haven't install 'remotes' package before 
remotes::install_local(path = "C:/Users/HandyCNV-1.1.7.zip") # Repalce 'C:/Users/' to your local path where you downloaded the Source Code</a></code></pre>
<p>Then, we need to load the ‘HandyCNV’ package in order to run the following examples. This can be done using the <code>library</code> function as shown below. <code>{r setup} library(HandyCNV)</code></p>
</div>
</div>
<div id="what-issues-can-this-package-solve" class="section level1">
<h1 class="hasAnchor">
<a href="#what-issues-can-this-package-solve" class="anchor"></a>What issues can this package solve?</h1>
<p>Click the following link to browse the output in examples.</p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#1-how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary-">1. How do we prepare the standard cnv input file for HandyCNV and get a quick summary?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#1-how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary-">2. How do we visualise CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#3-what-types-of-summary-plots-are-available-">3. What types of summary plots are available?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#4-how-do-we-generate-cnvrs-cnv-regions-from-cnv-results-">4. How do we generate CNVRs (CNV Regions) from CNV results?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#5-how-do-we-annotate-genes-for-cnvrs-and-cnvs-">5. How do we annotate genes for CNVRs and CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#6-how-to-plot-cnvr-distribution-map-">6. How to plot CNVR distribution map?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#7-how-can-we-plot-all-high-frequency-cnvrs-at-once-">7. How can we plot all high frequency CNVRs at once?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#8-can-we-compare-cnvs-between-different-result-sets-">8. Can we compare CNVs between different result sets?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#9-what-about-cnvrs-">9. What about CNVRs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#10-how-do-we-find-the-consensus-set-of-genes-common-to-multiple-cnv-result-sets-">10. How do we find the consensus set of genes common to multiple CNV result sets?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#11-how-can-we-create-the-map-file-used-in-handycnv-to-allow-comparing-cnvs-between-different-reference-genomes-">11. How can we create the map file used in HandyCNV, to allow comparing CNVs between different reference genomes?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#12-how-can-we-integrate-the-cnvs-and-annotated-genes-with-additional-information-such-as-log-r-ratio-b-allele-frequency-call-rate-heterozygosity-missing-value-rate-and-linkage-disequilibrium-and-plot-it-as-one-figure-">12. How can we integrate the CNVs and annotated genes with additional information, such as Log R ratio, B Allele Frequency, call rate, heterozygosity, missing value rate and Linkage Disequilibrium, and plot it as one figure?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#13-how-can-we-make-a-plot-to-show-the-source-of-the-cnvs-">13. How can we make a plot to show the source of the CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#14-how-can-we-plot-just-the-genes-in-a-specific-region-to-save-as-a-seperate-figure-">14. How can we plot just the genes in a specific region, to save as a seperate figure?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#15-how-can-we-find-regions-with-high-frequencies-of-runs-of-homozygosity-">15. How can we find regions with high frequencies of runs of homozygosity?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#16-how-do-we-visualise-rohs-">16. How do we visualize ROHs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#17-how-do-we-get-haplotype-for-roh-region-">17. How do we get haplotype for ROH region?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#18-how-do-we-convert-coordinates-for-cnv-cnvr-roh-or-any-other-intervals-">18. How do we convert coordinates for CNV, CNVR, ROH, or any other intervals?</a></p>
</div>
<div id="feature-request" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-request" class="anchor"></a>Feature request</h1>
<p>If you have any special requirements for this package, please feel free to sumbit your demands via this link: <a href="https://github.com/JH-Zhou/HandyCNV/issues/new?assignees=&amp;labels=&amp;template=feature-request.md">Submit Requirments</a>, we are happy to add the new features to meet your needs.</p>
</div>
<div id="bug-report" class="section level1">
<h1 class="hasAnchor">
<a href="#bug-report" class="anchor"></a>Bug report</h1>
<p>If you find any errors while using this package, please tell us via this link: <a href="https://github.com/JH-Zhou/HandyCNV/issues/new?assignees=&amp;labels=&amp;template=bug_report.md">Bug Report</a>, we will fix it as soon as possible.</p>
</div>
<div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>If this tool is useful for your academic research, please cite our publication: <a href="https://www.frontiersin.org/articles/10.3389/fgene.2021.731355/abstract">Browse publication</a></p>
<p>Citation: Zhou J, Liu L, Lopdell TJ, Garrick DJ and Shi Y (2021). HandyCNV: Standardized Summary, Annotation, Comparison and Visualization of CNV, CNVR, and ROH. Front. Genet. 12:731355. doi: 10.3389/fgene.2021.731355</p>
</div>
<div id="current-release-handycnv-v117-release-date-20210924" class="section level1">
<h1 class="hasAnchor">
<a href="#current-release-handycnv-v117-release-date-20210924" class="anchor"></a>Current release: HandyCNV v1.1.7 Release Date: 2021/09/24</h1>
</div>
<div id="whats-news" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news" class="anchor"></a>What’s news</h1>
<ol>
<li>New feature to visualize haplotype that generate from get_haplotype()</li>
</ol>
<pre class="{r}"><code><a href="reference/haplo_visual.html">haplo_visual(haplotype = haplotype_letter, xlab_text = "BMP7 Gene ")</a></code></pre>
<p><img src="https://github.com/JH-Zhou/HandyCNV/blob/master/vignettes/BMP7.jpg" alt="Fig. Haplotype of BMP7 Gene"></p>
<ol>
<li>Unify format of some inputs and outputs</li>
</ol>
</div>
<div id="current-release-handycnv-v116-release-date-20210901" class="section level1">
<h1 class="hasAnchor">
<a href="#current-release-handycnv-v116-release-date-20210901" class="anchor"></a>Current release: HandyCNV v1.1.6 Release Date: 2021/09/01</h1>
</div>
<div id="whats-news-1" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-1" class="anchor"></a>What’s news</h1>
<ol>
<li>Add autosomal boundary data for other species in ‘cnvr_plot’ function, which will be used to plot CNVR map. Now it support the hg38 and hg19 of human, UMD3.1 and ARS-UCD1.2 of cattle, Oar_v4.0 of sheep, Sscrofa11.1 of Pig, galGal6 of Chicken, EquCab3.0 of Horse and UMICH_Zoey_3.1 of Dog.</li>
<li>Add conditions to automatically control the number of X axis labels that present in the figure of The Number of CNVs Detected per Individual.</li>
<li>Update Github pages with Vignettes</li>
</ol>
</div>
<div id="previous-release-handycnv-v115-release-date-20210829" class="section level1">
<h1 class="hasAnchor">
<a href="#previous-release-handycnv-v115-release-date-20210829" class="anchor"></a>Previous release: HandyCNV v1.1.5 Release Date: 2021/08/29</h1>
</div>
<div id="whats-news-2" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-2" class="anchor"></a>What’s news</h1>
<p>Minor modifications, such as unifying input file formats and correcting spelling errors.</p>
</div>
<div id="previous-release-handycnv-v114-release-date-20210723" class="section level1">
<h1 class="hasAnchor">
<a href="#previous-release-handycnv-v114-release-date-20210723" class="anchor"></a>Previous release: HandyCNV v1.1.4 Release Date: 2021/07/23</h1>
</div>
<div id="whats-news-3" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-3" class="anchor"></a>What’s news</h1>
<div id="major-improvements" class="section level2">
<h2 class="hasAnchor">
<a href="#major-improvements" class="anchor"></a>Major improvements:</h2>
<ol>
<li><p>Most functions now support reading variable object as input files;</p></li>
<li><p>Most functions now support returning the main output as object to R environment for the further operation;</p></li>
<li>
<p>New function ‘get_samples’ to extract samples ID by searching interested gene from CNV annotation list.</p>
<h2 id="minor-changes">Minor changes:</h2>
</li>
<li><p>The ‘call_cnvr’ funtion now support generating CNVRs from CNV list that contains Chromosomes without CNVs information;</p></li>
<li><p>Add links of Horse_quCab2.0 genome reference and sheep ‘oviAri3’ reference genome into ‘get_refgene’ function;</p></li>
<li><p>Setup a standard table to support present comparison plot with empty group in ‘compare_cnvr’ function;</p></li>
<li><p>Add ‘-’ as separator between the two recoded haplotypes in ‘get_haplotype’ function.</p></li>
</ol>
</div>
</div>
<div id="previous-release-handycnv-v113-release-date-20210526" class="section level1">
<h1 class="hasAnchor">
<a href="#previous-release-handycnv-v113-release-date-20210526" class="anchor"></a>Previous release: HandyCNV v1.1.3 Release Date: 2021/05/26</h1>
</div>
<div id="whats-news-4" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-4" class="anchor"></a>What’s news</h1>
<ol>
<li><p>New function to plot SNP density from SNP genotyping map. <code>{r, warning=FALSE} plot_snp_density(map = "convert_map/target_plink.map",                   max_chr = 24, #optional                  top_density = 60, #optional                   low_density = 20, #optional                  color_top = "red", #optional                  color_low = "blue", #optional                  color_mid = "black", #optional                  legend_position = c(0.9, 0.1), #optional                   x_label = "Physical position\n物理位置", #optional                  y_label = "SNPs/Mb\n每1Mb区间的SNP数",#optional                  ncol_1 = 5)  #save the plot by 'ggsave' #ggsave(filename = "snp_density.png", width = 26, height = 18, units = "cm", dpi = 350)</code> <img src="https://github.com/JH-Zhou/HandyCNV/blob/master/vignettes/snp_density.png" alt="Fig.Demo SNP density distribution"></p></li>
<li><p>Revised CNVs status distribution plot in ‘cnv_summary_plot’ function, force to appear the boxplot and line on chromosome that has no CNVs.</p></li>
</ol>
</div>
<div id="previous-release-handycnv-v112-release-date-20210418" class="section level1">
<h1 class="hasAnchor">
<a href="#previous-release-handycnv-v112-release-date-20210418" class="anchor"></a>Previous release: HandyCNV v1.1.2 Release Date: 2021/04/18</h1>
</div>
<div id="whats-news-5" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-5" class="anchor"></a>What’s news</h1>
<p>Corrected the version number.</p>
</div>
<div id="released-version-handycnv-v111-release-date-20210414" class="section level1">
<h1 class="hasAnchor">
<a href="#released-version-handycnv-v111-release-date-20210414" class="anchor"></a>Released Version: HandyCNV v1.1.1 Release Date: 2021/04/14</h1>
</div>
<div id="whats-news-6" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-news-6" class="anchor"></a>What’s news</h1>
<div id="1-update-call_cnvrr-" class="section level2">
<h2 class="hasAnchor">
<a href="#1-update-call_cnvrr-" class="anchor"></a>1. Update call_cnvr.R …</h2>
<ol>
<li>The CNV list could only be loaded from the local directory through a ‘Path’ before, now supports to read data from working environment by checking the type of input file;</li>
<li>Support to return the CNVR list to working environment.</li>
</ol>
</div>
<div id="2-update-cnv_cleanr-" class="section level2">
<h2 class="hasAnchor">
<a href="#2-update-cnv_cleanr-" class="anchor"></a>2. Update cnv_clean.R …</h2>
<ol>
<li>Support to return Clean CNV List to working environment.</li>
</ol>
</div>
<div id="3-update-cnv_visualisingr-" class="section level2">
<h2 class="hasAnchor">
<a href="#3-update-cnv_visualisingr-" class="anchor"></a>3. Update cnv_visualising.R …</h2>
<ol>
<li>Support to load CNV List from working environment.</li>
</ol>
</div>
<div id="4-update-compare_cnvrr-" class="section level2">
<h2 class="hasAnchor">
<a href="#4-update-compare_cnvrr-" class="anchor"></a>4. Update compare_cnvr.R …</h2>
<ol>
<li>Add new function to generate the Unique and Mutual CNVRs by uniting the overlapped CNVRs between two results. There are two purposes of this work, one is to better understand the overlapping CNVRs, the other is to mark the common regions on CNVR distribution map in ‘cnvr_plot’ function;</li>
</ol>
</div>
<div id="5-update-cnvr_plot-" class="section level2">
<h2 class="hasAnchor">
<a href="#5-update-cnvr_plot-" class="anchor"></a>5. Update ‘cnvr_plot’ …</h2>
<ol>
<li>Add ‘overlap_cnvr’ argument to support to mark overlapped region on CNVR distribution map;</li>
<li>Add ‘label_prop’ argument to show the proportion of CNVRs length to total length of relative chromosome on CNVR map;</li>
<li>Add ‘chr_col’ argument to customize the color of Chromosome;</li>
<li>Add ‘overlap_col’ argument to customize color of overlapped CNVRs</li>
<li>Reduce the margin of final CNVR distribution map.</li>
</ol>
<div id="new-feature-demo" class="section level3">
<h3 class="hasAnchor">
<a href="#new-feature-demo" class="anchor"></a>New feature demo</h3>
<pre class="{r}"><code><a href="reference/cnvr_plot.html"># Demo code:
cnvr_plot(cnvr = "./cnvr_combine_part_penn_umd/cnvr.txt", assembly = "UMD", 
          sample_size =  388, common_cnv_threshold = 0.05, 
          overlap_cnvr = "./compare_cnvr_Penn_UMD_Vs_Part_UMD/common_cnvr.txt", 
          gain_col = "deeppink1", loss_col = "deepskyblue3", mixed_col = "springgreen3", 
          folder = "./cnvr_combine_part_penn_umd/cnvr_map_common")</a></code></pre>
<p><img src="https://github.com/JH-Zhou/HandyCNV/blob/master/vignettes/cnvr_plot.png" alt="Fig.2 CNVR Map"></p>
</div>
</div>
<div id="6-update-compare_gener-" class="section level2">
<h2 class="hasAnchor">
<a href="#6-update-compare_gener-" class="anchor"></a>6. Update compare_gene.R …</h2>
<ol>
<li>Add ‘color_label’ argument to display the color of genes that passed common threshold in the Heatmap;</li>
</ol>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jinghang Zhou <br><small class="roles"> Author, maintainer, copyright holder </small>  </li>
<li>Liyuan Liu <br><small class="roles"> Author </small>  </li>
<li>Thomas Lopdell <br><small class="roles"> Author </small>  </li>
<li>Dorian Garrick <br><small class="roles"> Author, thesis advisor </small>  </li>
<li>Yuangang Shi <br><small class="roles"> Author, thesis advisor </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JH-Zhou/HandyCNV/actions/workflows/R-CMD-check.yaml"><img src="https://github.com/JH-Zhou/HandyCNV/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/JH-Zhou/HandyCNV"><img src="https://codecov.io/gh/JH-Zhou/HandyCNV/branch/master/graph/badge.svg?token=5ERXZH0GFW" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Jinghang Zhou, Liyuan Liu, Thomas Lopdell, Dorian Garrick, Yuangang Shi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
