<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HandyCNV: An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH  • HandyCNV</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="HandyCNV: An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH ">
<meta property="og:description" content="The package is aimed to provide a Standard, Reproducible and Time-saving pipelines for the post-analysis of CNV, CNVR and ROH.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">HandyCNV</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/HandyCNV.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="welcome-to-handycnv" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#welcome-to-handycnv" class="anchor"></a>Welcome to HandyCNV</h1></div>

<p>An R package for Standardized Summary, Annotation, Comparison, and Visualization of CNV, CNVR and ROH</p>
<p><img src="https://github.com/JH-Zhou/HandyCNV/blob/master/man/figures/High_resolution_pipeline_HandyCNV.png" alt="Fig.1 Working Flow"></p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package was originally designed for the Post-analysis of CNV results inferred from PennCNV and CNVPartition (GenomeStudio). However, it has now been expanded to accept input files in standard formats for a wider range of applications. Our motivation is to provide a standard, reproducible and time-saving pipeline for the post-analysis of CNVs and ROHs detected from SNP genotyping data for the majority of diploid Species. The functions provided in this package can be categorised into five sections: Conversion, Summary, Annotation, Comparison and Visualization. The most useful features provided are: integrating summarized results, generating lists of CNVR, annotating the results with known gene positions, plotting CNVR distribution maps, and producing customised visualisations of CNVs and ROHs with gene and other related information on one plot. This package also supports a range of customisations, including the colour, size of high resolution figures, and output folder, avoiding conflict between the results of different runs. Running through all functions detailed in the vignette could help us to identify and explore the most interesting genomic regions more easily.</p>
</div>
<div id="vignettes-and-manual" class="section level1">
<h1 class="hasAnchor">
<a href="#vignettes-and-manual" class="anchor"></a>Vignettes and Manual</h1>
<p>The details examples please visit our Github pages: <a href="https://jh-zhou.github.io/HandyCNV/" class="uri">https://jh-zhou.github.io/HandyCNV/</a></p>
</div>
<div id="installation-and-prerequisites" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-prerequisites" class="anchor"></a>Installation and Prerequisites</h1>
<p>First, to run this package, we need to make sure that R (Version &gt;= 3.5.2) is installed in your computer (R download link: <a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a>). Once R is installed, the ‘HandyCNV’ package can be installed from Github repository by running the following script. If you rarely used R, it may take more time to install the ‘HandyCNV’ for the first time.</p>
<pre class="{r}"><code>#install.packages("remotes")
#library(remotes)
#install_github(repo = "JH-Zhou/HandyCNV", auth_token = "3d2ac98e4c297bab332f1e68b3b2d49f3a17d6aa")</code></pre>
<p>Then, we need to load the ‘HandyCNV’ package in order to run the following examples. This can be done using the <code>library</code> function as shown below. <code>{r setup} library(HandyCNV)</code></p>
</div>
<div id="what-issues-can-this-package-solve" class="section level1">
<h1 class="hasAnchor">
<a href="#what-issues-can-this-package-solve" class="anchor"></a>What issues can this package solve?</h1>
<p>Click the following link to browse the output in examples.</p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#1-how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary-">1. How do we prepare the standard cnv input file for HandyCNV and get a quick summary?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#1-how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary-">2. How do we visualise CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#3-what-types-of-summary-plots-are-available-">3. What types of summary plots are available?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#4-how-do-we-generate-cnvrs-cnv-regions-from-cnv-results-">4. How do we generate CNVRs (CNV Regions) from CNV results?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#5-how-do-we-annotate-genes-for-cnvrs-and-cnvs-">5. How do we annotate genes for CNVRs and CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#6-how-to-plot-cnvr-distribution-map-">6. How to plot CNVR distribution map?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#7-how-can-we-plot-all-high-frequency-cnvrs-at-once-">7. How can we plot all high frequency CNVRs at once?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#8-can-we-compare-cnvs-between-different-result-sets-">8. Can we compare CNVs between different result sets?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#9-what-about-cnvrs-">9. What about CNVRs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#10-how-do-we-find-the-consensus-set-of-genes-common-to-multiple-cnv-result-sets-">10. How do we find the consensus set of genes common to multiple CNV result sets?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#11-how-can-we-create-the-map-file-used-in-handycnv-to-allow-comparing-cnvs-between-different-reference-genomes-">11. How can we create the map file used in HandyCNV, to allow comparing CNVs between different reference genomes?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#12-how-can-we-integrate-the-cnvs-and-annotated-genes-with-additional-information-such-as-log-r-ratio-b-allele-frequency-call-rate-heterozygosity-missing-value-rate-and-linkage-disequilibrium-and-plot-it-as-one-figure-">12. How can we integrate the CNVs and annotated genes with additional information, such as Log R ratio, B Allele Frequency, call rate, heterozygosity, missing value rate and Linkage Disequilibrium, and plot it as one figure?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#13-how-can-we-make-a-plot-to-show-the-source-of-the-cnvs-">13. How can we make a plot to show the source of the CNVs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#14-how-can-we-plot-just-the-genes-in-a-specific-region-to-save-as-a-seperate-figure-">14. How can we plot just the genes in a specific region, to save as a seperate figure?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#15-how-can-we-find-regions-with-high-frequencies-of-runs-of-homozygosity-">15. How can we find regions with high frequencies of runs of homozygosity?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#16-how-do-we-visualise-rohs-">16. How do we visualize ROHs?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#17-how-do-we-get-haplotype-for-roh-region-">17. How do we get haplotype for ROH region?</a></p>
<p><a href="https://jh-zhou.github.io/HandyCNV/articles/HandyCNV.html#18-how-do-we-convert-coordinates-for-cnv-cnvr-roh-or-any-other-intervals-">18. How do we convert coordinates for CNV, CNVR, ROH, or any other intervals?</a></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>JH Zhou <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JH-Zhou/HandyCNV/actions/workflows/R-CMD-check.yaml"><img src="https://github.com/JH-Zhou/HandyCNV/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/JH-Zhou/HandyCNV"><img src="https://codecov.io/gh/JH-Zhou/HandyCNV/branch/master/graph/badge.svg?token=5ERXZH0GFW" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by JH Zhou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
