<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HandyCNV Vignette • HandyCNV</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="HandyCNV Vignette">
<meta property="og:description" content="HandyCNV">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HandyCNV</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/HandyCNV.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="HandyCNV_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>HandyCNV Vignette</h1>
                        <h4 class="author">Jinghang ZHOU, Thomas Lopdell, Liyuan Liu, Dorian Garrick, Yuangang Shi</h4>
            
            <h4 class="date">July 23, 2021</h4>
      
      
      <div class="hidden name"><code>HandyCNV.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package was originally designed for the Post-analysis of CNV results inferred from PennCNV and CNVPartition (GenomeStudio). However, it has now been expanded to accept input files in standard formats for a wider range of applications. Our motivation is to provide a standard, reproducible and time-saving pipeline for the post-analysis of CNVs and ROHs detected from SNP genotyping data for the majority of diploid Species. The functions provided in this package can be categorised into five sections: Conversion, Summary, Annotation, Comparison and Visualization. The most useful features provided are: integrating summarized results, generating lists of CNVR, annotating the results with known gene positions, plotting CNVR distribution maps, and producing customised visualisations of CNVs and ROHs with gene and other related information on one plot. This package also supports a range of customisations, including the colour, size of high resolution figures, and output folder, avoiding conflict between the results of different runs. Running through all functions detailed in the vignette could help us to identify and explore the most interesting genomic regions more easily. In the following sections, we will present how to use the functions provided in this package to solve these problems.</p>
<div id="installation-and-prerequisites" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-and-prerequisites" class="anchor"></a>Installation and Prerequisites</h2>
<p>First, to run this package, we need to make sure that R (Version &gt;= 3.5.2) is installed in your computer (R download link: <a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a>). Once R is installed, the ‘HandyCNV’ package can be installed from Github repository by running the following script. If you rarely used R, it may take more time to install the ‘HandyCNV’ for the first time.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#install.packages("remotes")</span>
<span class="co">#library(remotes)</span>
<span class="co">#install_github(repo = "JH-Zhou/HandyCNV")</span></code></pre></div>
<p>Then, we need to load the ‘HandyCNV’ package in order to run the following examples. This can be done using the <code>library</code> function as shown below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">HandyCNV</span><span class="op">)</span></code></pre></div>
<p>To start playing with this package, we first need to prepare at least one CNV result list. With only CNV results as an input file, we can explore the functions from section 1 to section 10 as below. But to get more interesting and potentially valuable results, we will have to prepare additional input files, including the map file and signal intensity file for the SNP chip used to generate the CNV list, the pedigree of the samples, and Plink format (bed, bim and fam) genotype files. Many of these files are already required during CNV discovery, and by generating them early in the pipeline, you will find that the rest of workflow will be much easier and faster. We will be introducing how to prepare each of the input files in the relevant chunks below.</p>
</div>
<div id="demo-data" class="section level2">
<h2 class="hasAnchor">
<a href="#demo-data" class="anchor"></a>Demo data</h2>
<p>We have provided some internal demo data, which should installed with package, in order to demonstrate how to use this package. We can access the demo data using the built-in R function <code>system.file</code> (see examples below). In the demo data, the first CNV results file is the default output from PennCNV with the ARS-UCD 1.2 (ARS) cattle reference genome. The second CNV results file contains the default output from CNVPartition with UMD 3.1 (UMD) cattle reference genome, and the third CNV list is an example of a standard input file that could be prepared from CNV results inferred by other tools.<br></p>
<p>Note: If you prefer not to test with this demo data, just skip this step to section 1 and use your own data instead.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnv_penn_ars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/cnv_results/ARS_PennCNV_WGC.goodcnv"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span>
<span class="va">cnv_part_umd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/cnv_results/UMD_CNVPartition.txt"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span>
<span class="va">cnv_other_umd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/cnv_results/UMD_Standard_CNV.txt"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span></code></pre></div>
</div>
<div id="setting-up-a-working-directory" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-a-working-directory" class="anchor"></a>Setting up a working directory</h2>
<p>To start the analysis, let’s first set up the working directory. This will help ensure that all results will be saved in the same directory.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="st">"C:/Users/handy_test"</span><span class="op">)</span> <span class="co">#remember to replace the path with your own </span></code></pre></div>
</div>
</div>
<div id="what-issues-can-this-package-solve" class="section level1">
<h1 class="hasAnchor">
<a href="#what-issues-can-this-package-solve" class="anchor"></a>What issues can this package solve?</h1>
<p>As above we mentioned, the functions in the package have been categorised into five sections: Conversion, Summary, Annotation, Comparison and Visualization.</p>
<p>Functions in the Conversion section can, for example, be used to convert coordinates between the default and target mapfiles provided by users; this would be useful if we are detecting CNVs with two different reference genomes. The converted map files can be output in formats suitable for use in Plink, PennCNV and our package ‘HandyCNV’. It is also possible to convert coordinates for intervals (CNV, CNVR, QTL et al) between the defaults and target maps, which might useful when comparing the results to those of other researchers. Functions included in the Summarization section allow the formatting and plotting of outputs, such as producing detailed summaries of CNVs and ROHs, making CNV summary plots, generating CNV regions (CNVRs), and plotting CNVR distribution maps.</p>
<p>The Annotation section allows the downloading of reference gene lists from the UCSC website, enabling the annotation of CNV, CNVR, ROH or other intervals with gene locations. The Comparison section facilitates the comparison of CNVs, CNVRs, Gene Frequent Lists and any other intervals. The highlight of this section is that comparison of CNVs and CNVRs will produce reports detailing comparison results on both individual and population levels, and all comparison results are relative to both input files. Finally, the Visualization section contains functions to support customised CNV and ROH plotting by chromosome, specific sample, regions of interest, or target genes (plotting by gene is not yet available for ROH). It is also possible to annotate the plots with other information, such as gene locations, log R ratio, B Allele Frequency, SNP genotype, LD or the source of CNVs. These functions can also plot genes separately from reference gene lists, which might useful for comparing outputs with plots from other studies.</p>
<p>Now, let’s start exploring the functions in <code>HandyCNV</code>.</p>
<div id="how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-prepare-the-standard-cnv-input-file-for-handycnv-and-get-a-quick-summary" class="anchor"></a>1. How do we prepare the standard cnv input file for <code>HandyCNV</code> and get a quick summary?</h2>
<p>We can use <code>cnv_clean</code> function to convert cnv results to standard format. This function will also produce a brief summary of CNV results. First, let’s format the sample CNV results from PennCNV. Note:</p>
<ol style="list-style-type: lower-alpha">
<li>When formatting results in PennCNV format, the corresponding CNV input argument is ‘penncnv’.</li>
<li>Default PennCNV results in Sample ID columns usually have the file path ahead the Sample ID. For example with the file “/home/zjhang027/C/final_403_penncnv/201094560082_R03C01.txt”, the pattern ‘cnv/’ before the sample ID is unique, so we can take it as the separator using the argument <code>penn_id_sep = "cnv/"</code> to extract the wanted sample ID.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnv_penn_ars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cnv_clean.html">cnv_clean</a></span><span class="op">(</span>penncnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#replace input file when running your own data</span>
                 penn_id_sep <span class="op">=</span> <span class="st">"cnv/"</span>, <span class="co">#take the unique pattern as a separator to extract the sample ID </span>
                 drop_length <span class="op">=</span> <span class="fl">5</span>, <span class="co"># the maximum CNV length threshold (in Mb); CNVs larger than this value will be deleted</span>
                 folder <span class="op">=</span> <span class="st">"cnv_clean"</span><span class="op">)</span> <span class="co">#create a new folder to save results </span>
<span class="co">#&gt; There are 382 individuals with 5215 CNVs in total.</span>
<span class="co">#&gt; The average number of CNVs present per individual is 13.65</span>
<span class="co">#&gt; Basic summary stats by CNV type:</span>
<span class="co">#&gt; # A tibble: 4 x 5</span>
<span class="co">#&gt;   CNV_Value     N `Average Length` `Min Length` `Max Length`</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1 0           146            65930         9932       343728</span>
<span class="co">#&gt; 2 1          2446           129081         1181      2822505</span>
<span class="co">#&gt; 3 3          2450           257261         6735      3651356</span>
<span class="co">#&gt; 4 4           173           687524        17406      4917181</span></code></pre></div>
<p>Next, let’s format a second set of CNV results, this time produced using CNVPartition. Note: The corresponding input argument is <code>cnvpartition</code> for CNV result from CNVPartition.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnv_part_umd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cnv_clean.html">cnv_clean</a></span><span class="op">(</span>cnvpartition <span class="op">=</span> <span class="va">cnv_part_umd</span>, <span class="co">#when running your own data, set this argument by 'cnvpartition = "localpath/file"' </span>
                 drop_length <span class="op">=</span> <span class="fl">5</span>, <span class="co">#the maximum CNV length threshold (in Mb); CNVs larger than this value will be deleted</span>
                 folder <span class="op">=</span> <span class="st">"cnv_clean"</span><span class="op">)</span> 
<span class="co">#&gt; There are 365 individuals with 1239 CNVs in total.</span>
<span class="co">#&gt; The average number of CNV on each Individual is 3.39</span>
<span class="co">#&gt; Basic summary stats by CNV type:</span>
<span class="co">#&gt; # A tibble: 4 x 5</span>
<span class="co">#&gt;   CNV_Value     N `Average Length` `Min Length` `Max Length`</span>
<span class="co">#&gt;       &lt;int&gt; &lt;int&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1         0   754           148699         4510      1204452</span>
<span class="co">#&gt; 2         1    59           375120        14229      3440849</span>
<span class="co">#&gt; 3         3   409           840778         4908      4867751</span>
<span class="co">#&gt; 4         4    17           634036        58524      2336934</span></code></pre></div>
<p>If the CNV results were not produced by PennCNV or CNVPartition, we need to arrange the CNV results into the standard format, then use the <code>standard_cnv</code> argument instead. The standard CNV input list should include the following columns (in order): <code>Sample_ID</code>, <code>Chr</code>, <code>Start</code>, <code>End</code> and <code>CNV_Value</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'knitr' was built under R version 3.5.3</span>
<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Sample_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"201977910077_R05C01"</span>, <span class="st">"201094560084_R06C02"</span>, <span class="st">"201094560084_R06C02"</span>, <span class="st">"201094560067_R08C02"</span>, <span class="st">"201094560067_R08C02"</span><span class="op">)</span>,
           Chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">7</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span>,
           Start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9087392</span>, <span class="fl">28583247</span>, <span class="fl">37214745</span>, <span class="fl">6248795</span>, <span class="fl">66567441</span><span class="op">)</span>,
           End <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9250902</span>, <span class="fl">28621254</span>, <span class="fl">37238472</span>, <span class="fl">6425277</span>, <span class="fl">66830563</span><span class="op">)</span>,
           CNV_Value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Sample_ID</th>
<th align="right">Chr</th>
<th align="right">Start</th>
<th align="right">End</th>
<th align="right">CNV_Value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">201977910077_R05C01</td>
<td align="right">6</td>
<td align="right">9087392</td>
<td align="right">9250902</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">201094560084_R06C02</td>
<td align="right">7</td>
<td align="right">28583247</td>
<td align="right">28621254</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">201094560084_R06C02</td>
<td align="right">7</td>
<td align="right">37214745</td>
<td align="right">37238472</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">201094560067_R08C02</td>
<td align="right">4</td>
<td align="right">6248795</td>
<td align="right">6425277</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">201094560067_R08C02</td>
<td align="right">4</td>
<td align="right">66567441</td>
<td align="right">66830563</td>
<td align="right">3</td>
</tr>
</tbody>
</table>
<p>Let’s now run <code>cnv_clean</code> on the third Standard CNV input file. Note that we are using the argument <code>standard_cnv</code> for the third CNV result.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnv_standard</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cnv_clean.html">cnv_clean</a></span><span class="op">(</span>standard_cnv <span class="op">=</span> <span class="va">cnv_other_umd</span>, <span class="co">#when running your own data, set this argument by 'standard_cnv = "localpath/file"'</span>
                 drop_length <span class="op">=</span> <span class="fl">5</span>, <span class="co">#the maximum CNV length threshold (in Mb); CNVs larger than this value will be deleted</span>
                 folder <span class="op">=</span> <span class="st">"cnv_clean"</span><span class="op">)</span>
<span class="co">#&gt; There are 382 individuals with 4534 CNVs in total.</span>
<span class="co">#&gt; The average number of CNV per individual is 11.87</span>
<span class="co">#&gt; Basic summary stats by CNV type:</span>
<span class="co">#&gt; # A tibble: 4 x 5</span>
<span class="co">#&gt;   CNV_Value     N `Average Length` `Min Length` `Max Length`</span>
<span class="co">#&gt;       &lt;int&gt; &lt;int&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1         0    77            60964         9932       343900</span>
<span class="co">#&gt; 2         1  2187           138295            1      2843668</span>
<span class="co">#&gt; 3         3  2115           258623            1      3933681</span>
<span class="co">#&gt; 4         4   155           779553        17392      4869255</span></code></pre></div>
</div>
<div id="how-do-we-visualise-cnvs" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-visualise-cnvs" class="anchor"></a>2 How do we visualise CNVs?</h2>
<p>The <code>cnv_visual</code> function supports visualising CNVs by chromosome, single sample, regions of interest, and target genes.</p>
<div id="how-do-we-have-a-quick-overview-of-all-cnvs-on-each-chromosome" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-have-a-quick-overview-of-all-cnvs-on-each-chromosome" class="anchor"></a>2.1 How do we have a quick overview of all CNVs on each chromosome?</h3>
<p>Once we got the standard <code>clean_cnv</code> result, we can use the <code>cnv_visual</code> function to plot all CNVs, to have a quick look if there are any notable patterns in CNV results.</p>
<p>First, let’s visualise the whole PennCNV dataset across all chromosomes:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
           max_chr <span class="op">=</span> <span class="fl">29</span>, <span class="co">#select how many chromosomes to plot </span>
           width_1 <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional,adjust the width of final plot</span>
           height_1 <span class="op">=</span><span class="fl">12</span>, <span class="co">#optional,adjust the height of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual_all_penn"</span><span class="op">)</span>
<span class="co">#&gt; Input data passed requirment check...</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 1"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 2"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 3"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 4"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 5"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 6"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 7"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 8"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 9"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 10"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 11"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 12"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 13"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 14"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 15"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 16"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 17"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 18"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 19"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 20"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 21"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 22"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 23"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 24"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 25"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 26"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 27"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 28"</span>
<span class="co">#&gt; [1] "Plotting the CNVs on Chromosome 29"</span>
<span class="co">#&gt; Task done.</span></code></pre></div>
<div class="figure">
<img src="cnv_visual_all_penn/Chr15.png" width="320" alt=""><p class="caption">Figure.1 CNVs on Chromosome 15 (ARS reference) from PennCNV</p>
</div>
<p>Let’s now visualise CNVs from the CNVPartition results by chromosome. We can adjust the size of final figure by setting ‘width_1’ and ‘height_1’.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_part_umd</span>, <span class="co">#generated by 'cnv_clean' function in section 1</span>
           max_chr <span class="op">=</span> <span class="fl">29</span>, <span class="co">#select how many Chromosomes to plot</span>
           width_1 <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional,adjust the width of final plot</span>
           height_1 <span class="op">=</span> <span class="fl">12</span>, <span class="co">#optional,adjust the height of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual_all_part"</span>, <span class="op">)</span></code></pre></div>
<div class="figure">
<img src="cnv_visual_all_part/Chr28.png" width="320" alt=""><p class="caption">Figure.2 CNVs on Chromosome 28 (UMD reference) from CNVPartition</p>
</div>
</div>
<div id="how-do-we-plot-cnvs-in-a-region-of-interest" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-plot-cnvs-in-a-region-of-interest" class="anchor"></a>2.2 How do we plot CNVs in a region of interest?</h3>
<p>After plotting all CNVs on each chromosome, we might find that some regions are particularly interesting. To plot these, we can add the options <code>chr_id</code>, <code>start_position</code> and <code>end_position</code> to the <code>cnv_visual</code> function call. Here, we are taking the region of Chr15:77-80.1221 Mb from the PennCNV results as an example.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
           chr_id <span class="op">=</span> <span class="fl">15</span>, <span class="co">#set Chromosome</span>
           start_position <span class="op">=</span> <span class="fl">77</span>, <span class="co">#set start position, unit is 'Mb'</span>
           end_position <span class="op">=</span> <span class="fl">80.1221</span>, <span class="co">#set end position, unit is 'Mb'</span>
           col_0 <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, customize colour for 0 copy CNVs </span>
           col_1 <span class="op">=</span> <span class="st">"orange"</span>, <span class="co">#optional, customize colour for 1 copy CNVs </span>
           col_3 <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional, customize colour for 3 copy CNVs </span>
           col_4 <span class="op">=</span> <span class="st">"green"</span>, <span class="co">#optional, customize colour for 4 copy CNVs </span>
           width_1 <span class="op">=</span> <span class="fl">13</span>, <span class="co">#optional, adjust the width of final plot</span>
           height_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#optional, adjust the height of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual"</span><span class="op">)</span>
<span class="co">#&gt; Input data passed requirment check...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnv_visual/Chr15_77-80.1221Mb.png" width="320" alt=""><p class="caption">Figure.3 CNVs on Chr15:77-80.1221 Mb</p>
</div>
</div>
<div id="how-do-we-plot-cnvs-in-a-sample-of-interest" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-plot-cnvs-in-a-sample-of-interest" class="anchor"></a>2.3 How do we plot CNVs in a sample of interest?</h3>
<p>Here we need to assign both the Sample ID (to <code>individual_id</code>) and the clean CNV list (to <code>clean_cnv</code>) in the <code>cnv_visual</code> function. The individual CNV plot will saved in the current working directory. For example, the sample “201094560076_R03C02” has the most CNVs in the PennCNV results, so let’s plot it and have a look. Note: we can use the <code>individual_cnv_count.txt</code> file (to be generated in section 3) to identify samples that exhibit large numbers of CNVs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
           individual_id <span class="op">=</span> <span class="st">"201094560076_R03C02"</span>, <span class="co">#set the sample ID  </span>
           width_1 <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional, adjust the width of final plot</span>
           height_1 <span class="op">=</span> <span class="fl">13</span>, <span class="co">#optional. adjust the height of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual"</span><span class="op">)</span>
<span class="co">#&gt; Input data passed requirment check...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnv_visual/CNV_of_201094560076_R03C02.png" width="320" alt=""><p class="caption">Figure.4 CNVs in a specific sample</p>
</div>
</div>
<div id="how-do-we-plot-an-interesting-gene-to-see-how-many-samples-had-a-cnv-overlapping-with-it" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-plot-an-interesting-gene-to-see-how-many-samples-had-a-cnv-overlapping-with-it" class="anchor"></a>2.4 How do we plot an interesting gene to see how many samples had a CNV overlapping with it?</h3>
<p>The idea of plotting CNVs by target gene is to avoid making overly-strong statements when reporting important candidate genes after annotation for CNVRs. If we think that one particular gene is especially important in the CNV analysis, it’s better to present the frequency at which that gene has a structural variation in the experimental population. Plotting CNVs by gene will directly present how many samples show copy number variations overlapping the gene.</p>
<p>Note: The input file ‘cnv_annotation.txt’ will be generated in section 5.2, as part of annotating the gene to CNV results.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="st">"call_gene_Penn_UCSC/cnv_annotation.txt"</span>, <span class="co">#here we need the annotated CNV list, generated by 'call_gene' in section 5.2</span>
           target_gene <span class="op">=</span> <span class="st">"BLA-DQB"</span>, <span class="co">#set the gene name we are interested in</span>
           col_0 <span class="op">=</span> <span class="st">"darkorchid4"</span>,  <span class="co">#optional, customize colour for o copy CNV </span>
           col_1 <span class="op">=</span> <span class="st">"dodgerblue3"</span>,  <span class="co">#optional, customize colour for 1 copy CNV </span>
           col_3 <span class="op">=</span> <span class="st">"violetred3"</span>,  <span class="co">#optional, customize colour for 3 copy CNV </span>
           col_4 <span class="op">=</span> <span class="st">"orangered3"</span>,  <span class="co">#optional, customize colour for 4 copy CNV</span>
           width_1 <span class="op">=</span> <span class="fl">13</span>, <span class="co">#adjust the width of final plot</span>
           height_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#adjust the width of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual"</span><span class="op">)</span>
<span class="co">#&gt; Renamed 'CNV_Start' and 'CNV_End' as 'Start' and 'End', input data passed requirment check...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnv_visual/BLA-DQB.png" width="320" alt=""><p class="caption">Figure.5 CNVs on a specific gene (<em>BLA-DQB</em>)</p>
</div>
</div>
<div id="how-can-we-present-annotated-genes-above-the-cnv-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#how-can-we-present-annotated-genes-above-the-cnv-plot" class="anchor"></a>2.5 How can we present annotated genes above the CNV plot?</h3>
<p>To do this, we need to provide two input files to the <code>cnv_visual</code> function: the CNV list in <code>clean_cnv</code> and gene annotations in <code>refgene</code>. We should also set the genomic region we are interested in. The reference gene list will be generated by the <code>get_refgene</code> function in section 5.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_visual.html">cnv_visual</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
           refgene <span class="op">=</span> <span class="st">"refgene/Cow_ARS_UCSC.txt"</span>, <span class="co">#standard input file will generated by the function 'get_refgene' in Section 5</span>
           chr_id <span class="op">=</span> <span class="fl">23</span>, <span class="co">#set Chromosome</span>
           start_position <span class="op">=</span> <span class="fl">25.56</span>, <span class="co">#set start position, unit is 'Mb'</span>
           end_position <span class="op">=</span> <span class="fl">25.74</span>, <span class="co">#set end position, unit is 'Mb'</span>
           gene_font_size <span class="op">=</span> <span class="fl">2.2</span>, <span class="co">#optional, adjust the size of gene font</span>
           show_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">25.66</span>,<span class="fl">27.72</span><span class="op">)</span>, <span class="co">#optional, show the gene in the given interval, maximum support two intervals</span>
           col_0 <span class="op">=</span> <span class="st">"tomato"</span>, <span class="co">#optional, customize colour for 0 copy CNV </span>
           col_1 <span class="op">=</span> <span class="st">"hotpink"</span>, <span class="co">#optional, customize colour for 1 copy CNV </span>
           col_3 <span class="op">=</span> <span class="st">"turquoise"</span>, <span class="co">#optional, customize colour for 3 copy CNV </span>
           col_4 <span class="op">=</span> <span class="st">"aquamarine"</span>, <span class="co">#optional, customize colour for 4 copy CNV </span>
           col_gene <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, customize colour of gene  </span>
           width_1 <span class="op">=</span> <span class="fl">14</span>, <span class="co">#optional, adjust the width of final plot</span>
           height_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#optional, adjust the height of final plot</span>
           folder <span class="op">=</span> <span class="st">"cnv_visual"</span><span class="op">)</span>
<span class="co">#&gt; Input data passed requirment check...</span>
<span class="co">#&gt; [1] "CNV on Chromosome 23: 25.56 - 25.74 Mb with 68 Individual - "</span>
<span class="co">#&gt; Plotting gene...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnv_visual/Chr23_25.56-25.74Mb.png" width="320" alt=""><p class="caption">Figure.6 CNVs with annotated genes</p>
</div>
</div>
</div>
<div id="what-types-of-summary-plots-are-available" class="section level2">
<h2 class="hasAnchor">
<a href="#what-types-of-summary-plots-are-available" class="anchor"></a>3. What types of summary plots are available?</h2>
<p>Using the clean CNV results, we can run the <code>cnv_summary_plot</code> function to generate a range of summary plots. These plots can aggregate CNV results by length group, CNV type, chromosome, and individual. It is also possible to make two different types of plots combining these figures.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnv_summary_plot.html">cnv_summary_plot</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
                 length_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.6</span>, <span class="fl">0.7</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span>, <span class="co"># optional, set group of vectors to divide CNV length, unit is Mb. such as vector of ‘c(0.05, 0.3, 1)’, means divide the CNV length into four group: '&lt;0.05Mb', '0.05 - 0.3Mb', '0.3-1Mb' and '&gt;1Mb', maximum can accept 11 values</span>es
                 col_0 <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, customize color for o copy CNV </span>
                 col_1 <span class="op">=</span> <span class="st">"black"</span>, <span class="co">#optional, customize color for o copy CNV </span>
                 col_3 <span class="op">=</span> <span class="st">"purple"</span>, <span class="co">#optional, customize color for o copy CNV </span>
                 col_4 <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional, customize color for o copy CNV </span>
                 plot_sum_1 <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co">#optional, make sum combination plot 1</span>
                 height_sum1 <span class="op">=</span> <span class="fl">26</span>, <span class="co">#optional, adjust the size of sum plot 1</span>
                 width_sum1 <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional, adjust the size of sum plot 1</span>
                 plot_sum_2 <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#optional, make sum combination plot 2</span>
                 height_sum2 <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional, adjust the size of sum plot 2</span>
                 width_sum2 <span class="op">=</span> <span class="fl">27</span>, <span class="co">#optional, adjust the size of sum plot 2</span>
                 folder <span class="op">=</span> <span class="st">"cnv_summary_plot_penn"</span><span class="op">)</span>
<span class="co">#&gt; Deviding the length group...</span>
<span class="co">#&gt; Plotting...</span>
<span class="co">#&gt; The CNV distribution plot has been saved in the 'cnv_summary_plot_penn' directory.</span>
<span class="co">#&gt; The CNV length group plot was saved in the 'cnv_summary_plot_penn' directory.</span>
<span class="co">#&gt; The box plot of CNV lengths was saved in the 'cnv_summary_plot_penn' directory.</span>
<span class="co">#&gt; Distribution of CNV type on chromosome plot was saved in working directory.</span>
<span class="co">#&gt; The number of CNV in Individual plot was saved in working directory.</span>
<span class="co">#&gt; cnv_summary_plot_1 was saved in working directory.</span>
<span class="co">#&gt; cnv_summaray_polt_2 was saved in working directory.</span>
<span class="co">#&gt; Task done.</span></code></pre></div>
<div class="figure">
<img src="cnv_summary_plot_Penn/cnv_summary_plot_1.png" width="320" alt=""><p class="caption">Figure.7 CNV Summary Plot Combination 1</p>
</div>
<div class="figure">
<img src="cnv_summary_plot_Penn/cnv_summaray_plot_2.png" width="320" alt=""><p class="caption">Figure.8 CNV Summary Plot Combination 2</p>
</div>
</div>
<div id="how-do-we-generate-cnvrs-cnv-regions-from-cnv-results" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-generate-cnvrs-cnv-regions-from-cnv-results" class="anchor"></a>4. How do we generate CNVRs (CNV Regions) from CNV results?</h2>
<p>CNV regions are defined as regions of the genome containing one or more CNVs that overlap by at least one base pair: s a result, CNVRs will never overlap one another. As an example, let’s first generate CNVRs for the clean PennCNV results, using the <code>call_cnvr</code> function. A key parameter of this function is <code>chr_set</code>, which allows setting the maximum number of chromosomes over which to generate CNVRs. In our example, the species is <em>Bos taurus</em> (cows), with 29 autosomes, so we set <code>chr_set = 29</code>. This function will output three tables:</p>
<ol style="list-style-type: decimal">
<li>The list of CNVRs. This is saved into the file <code>individual_cnv_cnvr.txt</code>.</li>
<li>A brief summary table showing numbers of CNVRs by length and type (Deletion, Duplication, and Mixed). Mixed indicates that both duplications and deletions are found within the CNVR.</li>
<li>The total length and number of CNVRs on each chromosome.</li>
</ol>
<p>The resulting CNVRs are sorted by chromosome, start position, and end position, then assigned the CNVR ID to each unique interval by order. One of the output files, <code>individual_cnv_cnvr.txt</code> contain both CNVR and CNV information for every sample: this file might suitable for CNV-GWAS analyses.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnvr_penn_ars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/call_cnvr.html">call_cnvr</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
                  chr_set <span class="op">=</span> <span class="fl">29</span>, <span class="co">#Set the maximum number of chromosomes (29 autosomes for cattle)</span>
                  folder <span class="op">=</span> <span class="st">"call_cnvr_penn"</span><span class="op">)</span>
<span class="co">#&gt; Chromosome 1 has been processed.</span>
<span class="co">#&gt; Chromosome 2 has been processed.</span>
<span class="co">#&gt; Chromosome 3 has been processed.</span>
<span class="co">#&gt; Chromosome 4 has been processed.</span>
<span class="co">#&gt; Chromosome 5 has been processed.</span>
<span class="co">#&gt; Chromosome 6 has been processed.</span>
<span class="co">#&gt; Chromosome 7 has been processed.</span>
<span class="co">#&gt; Chromosome 8 has been processed.</span>
<span class="co">#&gt; Chromosome 9 has been processed.</span>
<span class="co">#&gt; Chromosome 10 has been processed.</span>
<span class="co">#&gt; Chromosome 11 has been processed.</span>
<span class="co">#&gt; Chromosome 12 has been processed.</span>
<span class="co">#&gt; Chromosome 13 has been processed.</span>
<span class="co">#&gt; Chromosome 14 has been processed.</span>
<span class="co">#&gt; Chromosome 15 has been processed.</span>
<span class="co">#&gt; Chromosome 16 has been processed.</span>
<span class="co">#&gt; Chromosome 17 has been processed.</span>
<span class="co">#&gt; Chromosome 18 has been processed.</span>
<span class="co">#&gt; Chromosome 19 has been processed.</span>
<span class="co">#&gt; Chromosome 20 has been processed.</span>
<span class="co">#&gt; Chromosome 21 has been processed.</span>
<span class="co">#&gt; Chromosome 22 has been processed.</span>
<span class="co">#&gt; Chromosome 23 has been processed.</span>
<span class="co">#&gt; Chromosome 24 has been processed.</span>
<span class="co">#&gt; Chromosome 25 has been processed.</span>
<span class="co">#&gt; Chromosome 26 has been processed.</span>
<span class="co">#&gt; Chromosome 27 has been processed.</span>
<span class="co">#&gt; Chromosome 28 has been processed.</span>
<span class="co">#&gt; Chromosome 29 has been processed.</span>
<span class="co">#&gt; 978 CNVRs generated in total.</span>
<span class="co">#&gt; Overall summary of CNVRs:</span>
<span class="co">#&gt; # A tibble: 3 x 6</span>
<span class="co">#&gt;   Type      N `Average Length` `Min Length` `Max Length` `Total Length`</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Gain    347           150153         7947      2639041       52103038</span>
<span class="co">#&gt; 2 Loss    399           100664         3703       737404       40164889</span>
<span class="co">#&gt; 3 Mixed   232           898204         6735      8282577      208383407</span>
<span class="co">#&gt; Summary of CNVRs on each Chromosome:</span>
<span class="co">#&gt; # A tibble: 29 x 3</span>
<span class="co">#&gt;    Chr   `Total Length` `Number of CNVR`</span>
<span class="co">#&gt;    &lt;chr&gt;          &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt;  1 1           26386484               62</span>
<span class="co">#&gt;  2 10           6784489               35</span>
<span class="co">#&gt;  3 11          12434667               33</span>
<span class="co">#&gt;  4 12          23045567               34</span>
<span class="co">#&gt;  5 13           4035800               35</span>
<span class="co">#&gt;  6 14          11313259               31</span>
<span class="co">#&gt;  7 15          10540774               32</span>
<span class="co">#&gt;  8 16           9051167               31</span>
<span class="co">#&gt;  9 17          12211360               32</span>
<span class="co">#&gt; 10 18           3455220               35</span>
<span class="co">#&gt; # ... with 19 more rows</span></code></pre></div>
<p>Next, let’s generate CNVRs for CNVPartition and standard CNV lists. In order to avoid mixing the results among different files, we can set the <code>folder</code> argument to assign a distinguishing name to the newly generated folder.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/call_cnvr.html">call_cnvr</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_part_umd</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
                    chr_set <span class="op">=</span> <span class="st">'29'</span>, 
                    folder <span class="op">=</span> <span class="st">"call_cnvr_part"</span><span class="op">)</span>

<span class="fu"><a href="../reference/call_cnvr.html">call_cnvr</a></span><span class="op">(</span>clean_cnv <span class="op">=</span> <span class="va">cnv_standard</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
          chr_set <span class="op">=</span> <span class="fl">29</span>,
          folder <span class="op">=</span> <span class="st">"call_cnvr_standard"</span><span class="op">)</span></code></pre></div>
</div>
<div id="how-do-we-annotate-genes-for-cnvrs-and-cnvs" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-annotate-genes-for-cnvrs-and-cnvs" class="anchor"></a>5. How do we annotate genes for CNVRs and CNVs?</h2>
<p>Once we have defined some CNV regions, we might want to see how many genes are located each. We can annotate genes to both CNVs and CNVRs using the <code>call_gene</code> function. Before that, however, we need to prepare gene lists for the correct reference genome. This can be done using the <code>get_refgene</code> function, as described in the following sections.</p>
<div id="how-do-we-download-the-reference-gene-lists-we-need" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-download-the-reference-gene-lists-we-need" class="anchor"></a>5.1 How do we download the reference gene lists we need?</h3>
<p>We can use the <code>get_refgene</code> function to automatically download the reference gene list and convert it to the standard format required for functions such as <code>call_gene</code>. First, let’s run <code>get_refgene</code> without arguments to see which reference gene lists we can download. If you don’t find the information you’re looking for, please feel free to contact our maintainer to update it to you. For now, only reference gene lists from the UCSC website are supported; however, we may add more features if there is sufficient demand.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/get_refgene.html">get_refgene</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Please select which version to download from following list:</span>
<span class="co">#&gt;  [1] "Cow_ARS_UCSC"            "Cow_ARS_ENS"            </span>
<span class="co">#&gt;  [3] "Cow_UMD_UCSC"            "Pig_susScr11_UCSC"      </span>
<span class="co">#&gt;  [5] "Pig_susScr11_ENS"        "Human_hg38"             </span>
<span class="co">#&gt;  [7] "Human_hg19"              "Sheep_Oar_v4.0_UCSC"    </span>
<span class="co">#&gt;  [9] "Sheep_Oar_v3.1_UCSC"     "Sheep_Oar_v3.1_ENS"     </span>
<span class="co">#&gt; [11] "Horse_equCab3.0_UCSC"    "Horse_equCab2.0_UCSC"   </span>
<span class="co">#&gt; [13] "Dog_UMICH_Zoey_3.1_UCSC" "Dog_UMICH_Zoey_3.1_ENS" </span>
<span class="co">#&gt; [15] "Chicken_galGal6_UCSC"    "Chicken_galGal5_UCSC"</span></code></pre></div>
<p>Now, let’s get the reference gene list for the PennCNV results. This dataset was generated using ARS-UCD 1.2 as reference genome. Two gene lists are available for this reference: <code>cow_ARS_UCSC</code> and <code>cow_ARS_ENS</code>. The suffix <code>ENS</code> indicates the Ensembl reference gene list. The gene list will be saved in the folder <code>refgene</code>, which will be created first if necessary.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Cow_ARS_UCSC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_refgene.html">get_refgene</a></span><span class="op">(</span>gene_version <span class="op">=</span> <span class="st">"Cow_ARS_UCSC"</span><span class="op">)</span>
<span class="co">#&gt; Link to the website...</span>
<span class="co">#&gt; Converting format...</span>
<span class="co">#&gt; Task done.</span>
<span class="va">Cow_ARS_ENS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_refgene.html">get_refgene</a></span><span class="op">(</span>gene_version <span class="op">=</span> <span class="st">"Cow_ARS_ENS"</span><span class="op">)</span>
<span class="co">#&gt; Link to the website...</span>
<span class="co">#&gt; Converting format...</span>
<span class="co">#&gt; Task done.</span></code></pre></div>
<p>Next let’s get reference gene list for the CNVPartition and Standard result sets, both of which were generated on the UMD 3.1 reference genome; in this case, <code>Cow_UMD_UCSC</code> is the correct option to select.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">COW_UMD_UCSC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_refgene.html">get_refgene</a></span><span class="op">(</span>gene_version <span class="op">=</span> <span class="st">"Cow_UMD_UCSC"</span><span class="op">)</span>
<span class="co">#&gt; Link to the website...</span>
<span class="co">#&gt; Converting format...</span>
<span class="co">#&gt; Task done.</span></code></pre></div>
<p>Now we have well-prepared reference gene lists, let’s return to the <code>call_gene</code> function.</p>
</div>
<div id="the-call_gene-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-call_gene-function" class="anchor"></a>5.2 The <code>call_gene</code> function</h3>
<p>The <code>call_gene</code> function has three arguments:</p>
<ol style="list-style-type: decimal">
<li>
<code>ref_gene</code>: this argument is required, and takes the path of a downloaded gene set (in the <code>refgene</code> folder).</li>
<li>
<code>interval</code>: an optional file containing a list of CNV regions. This argument can also point to any file storing interval data in the correct format; examples could include CRVs, QTL windows, or haplotype lists.</li>
<li>
<code>clean_cnv</code>: an optional file containing a list of CNVs</li>
</ol>
<p>We suggest that both ‘interval’ and ‘clean_cnv’ should be set simultaneously: this will enable reporting of the frequency of all annotated genes by counting how many CNVs overlapped to each gene within a CNVR. In section 10, we will show how to use the output file <code>gene_freq_cnv.txt</code> to make comparisons between different CNV results to find sets of consensus genes.</p>
<p>As an example, we will first call genes for the PennCNV results, ensuring that the gene list for the correct reference has been set (ARS 1.2). In this case, we are using the Ensembl-provided gene list.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/call_gene.html">call_gene</a></span><span class="op">(</span>refgene <span class="op">=</span> <span class="st">"refgene/Cow_ARS_ENS.txt"</span>, <span class="co">#standard file was generated by 'get_refgene' function in section 5.1</span>
          interval <span class="op">=</span> <span class="va">cnvr_penn_ars</span>, <span class="co">#optional, standard file was generated by 'call_cnvr' function in section 4</span>
          clean_cnv <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#optional, standard file was generated by 'cnv_clean' function in section 1</span>
          folder <span class="op">=</span> <span class="st">"call_gene_Penn_Ens"</span><span class="op">)</span>
<span class="co">#&gt; Warning in call_gene(refgene = "refgene/Cow_ARS_ENS.txt", interval = cnvr_penn_ars, : The following genes are duplicated in multiple CNVRs, and will only be annotated on the first CNVR_ID in the final gene frequency list report!</span></code></pre></div>
<p>We can also use the UCSC gene list:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/call_gene.html">call_gene</a></span><span class="op">(</span>refgene <span class="op">=</span> <span class="st">"refgene/cow_ARS_UCSC.txt"</span>, <span class="co">#standard file was generated by 'get_refgene' function in section 5.1</span>
          interval <span class="op">=</span> <span class="st">"call_cnvr_Penn/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          clean_cnv <span class="op">=</span> <span class="st">"cnv_clean/penncnv_clean.cnv"</span>, <span class="co">#optional, standard file was generated by 'cnv_clean' function in section 1</span>
          folder <span class="op">=</span> <span class="st">"call_gene_Penn_UCSC"</span><span class="op">)</span>
<span class="co">#&gt; Warning in call_gene(refgene = "refgene/cow_ARS_UCSC.txt", interval = "call_cnvr_Penn/cnvr.txt", : The following genes are duplicated in multiple CNVRs, and will only be annotated on the first CNVR_ID in the final gene frequency list report!</span></code></pre></div>
<p>Comparing the results, we observe that the results from two reference gene lists are slightly different, likely because the maintainers of each database have different methodologies (and possibly data sources) for producing their gene annotations. As a result of this, the position and quantity of genes might slightly differ between gene builds even for the same reference genome. Therefore, it may be important to validate genes of interest in more than one database.</p>
<p>Now, let’s call genes for the CNVPartition results. As before, we provide both CNVR and CNV lists, and set the new folder name by arguments. Note that we need to use the UMD 3.1 reference build for this data set. We also run the standard CNV set (using the ARS1.2 reference).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/call_gene.html">call_gene</a></span><span class="op">(</span>refgene <span class="op">=</span> <span class="st">"refgene/cow_UMD_UCSC.txt"</span>, <span class="co">#standard file was generated by 'get_refgene' function in section 5.1</span>
          interval <span class="op">=</span> <span class="st">"call_cnvr_Part/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          clean_cnv <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_clean.cnv"</span>, <span class="co">#optional, standard file was generated by 'cnv_clean' function in section 1</span>
          folder <span class="op">=</span> <span class="st">"call_gene_Part"</span><span class="op">)</span>
<span class="co">#&gt; Warning in call_gene(refgene = "refgene/cow_UMD_UCSC.txt", interval = "call_cnvr_Part/cnvr.txt", : The following genes are duplicated in multiple CNVRs, and will only be annotated on the first CNVR_ID in the final gene frequency list report!</span>

<span class="fu"><a href="../reference/call_gene.html">call_gene</a></span><span class="op">(</span>refgene <span class="op">=</span> <span class="st">"refgene/cow_ARS_UCSC.txt"</span>, <span class="co">#standard file was generated by 'get_refgene' function in section 5.1</span>
          interval <span class="op">=</span> <span class="st">"call_cnvr_standard/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          clean_cnv <span class="op">=</span> <span class="st">"cnv_clean/cleancnv.cnv"</span>, <span class="co">#optional, standard file was generated by 'cnv_clean' function in section 1</span>
          folder <span class="op">=</span> <span class="st">"call_gene_standard_ARS_UCSC"</span><span class="op">)</span>
<span class="co">#&gt; Warning in call_gene(refgene = "refgene/cow_ARS_UCSC.txt", interval = "call_cnvr_standard/cnvr.txt", : The following genes are duplicated in multiple CNVRs, and will only be annotated on the first CNVR_ID in the final gene frequency list report!</span></code></pre></div>
</div>
</div>
<div id="how-to-plot-cnvr-distribution-map" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-plot-cnvr-distribution-map" class="anchor"></a>6. How to plot CNVR distribution map?</h2>
<p>The function <code>cnvr_plot</code> is used to plot the CNVR distribution across chromosomes. For this function, we only need to assign the a CNVR list file to the <code>cnvr</code> argument, and the reference genome assembly to the <code>assembly</code> argument (to provide chromosome lengths). For Bovine data, we provide data for both ARS and UMD. For other species with data available (see the <code>ref_gene</code> function in section 5.1), we can just assign the species name to the <code>assembly</code> argument, for example, <code>assembly = "human"</code> for human data.</p>
<p>First, we’ll plot the CNVR distribution for the PennCNV data set. The default legend position might not suitable for all species; the arguments <code>legend_x</code> and <code>legend_y</code> can be used to move it if necessary.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnvr_plot.html">cnvr_plot</a></span><span class="op">(</span>cnvr <span class="op">=</span> <span class="st">"call_cnvr_Penn/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          assembly <span class="op">=</span> <span class="st">"ARS"</span>, <span class="co">#cattle could select 'UMD' or 'ARS', other species could assign the name of species</span>
          loss_col <span class="op">=</span> <span class="st">"green2"</span>, <span class="co">#optional, adjust the color of Loss type of CNVR</span>
          gain_col <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, adjust the color of Gain type of CNVR</span>
          mixed_col <span class="op">=</span> <span class="st">"black"</span>,  <span class="co">#optional, adjust the color of Mixed type of CNVR</span>
          legend_x <span class="op">=</span> <span class="fl">127</span>, <span class="co">#optional, adjust the horizontal position of legend</span>
          legend_y <span class="op">=</span> <span class="fl">30</span>,  <span class="co">#optional, adjust the vertical position of legend</span>
          <span class="co">#height_1 = 14, #optional, adjust the height of CNVR plot</span>
          <span class="co">#width_1 = 10, #optional, adjust the width of CNVR plot</span>
          folder <span class="op">=</span> <span class="st">"cnvr_plot_penn"</span><span class="op">)</span>
<span class="co">#&gt; Task done, CNVR Distribution Plot saved in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnvr_plot_penn/cnvr_plot.png" width="320" alt=""><p class="caption">Figure.9 CNVR distribution map for PennCRV data</p>
</div>
<p>Next, we plot the CNVRs derived from the CNVPartition data set.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnvr_plot.html">cnvr_plot</a></span><span class="op">(</span>cnvr <span class="op">=</span> <span class="st">"call_cnvr_Part/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          assembly <span class="op">=</span> <span class="st">"UMD"</span>, <span class="co">#cattle could select 'UMD' or 'ARS', other species could assign the name of species</span>
          loss_col <span class="op">=</span> <span class="st">"darkorchid"</span>, <span class="co">#optional, adjust the color of Loss type of CNVR</span>
          gain_col <span class="op">=</span> <span class="st">"gold"</span>, <span class="co">#optional, adjust the color of Gain type of CNVR</span>
          mixed_col <span class="op">=</span> <span class="st">"tomato"</span>,  <span class="co">#optional, adjust the color of Mixed type of CNVR</span>
          legend_x <span class="op">=</span> <span class="fl">127</span>, <span class="co">#optional, adjust the horizontal position of legend</span>
          legend_y <span class="op">=</span> <span class="fl">30</span>,  <span class="co">#optional, adjust the vertical position of legend</span>
          height_1 <span class="op">=</span> <span class="fl">14</span>, <span class="co">#optional, adjust the height of CNVR plot</span>
          width_1 <span class="op">=</span> <span class="fl">17</span>, <span class="co">#optional, adjust the width of CNVR plot</span>
          folder <span class="op">=</span> <span class="st">"cnvr_plot_part"</span><span class="op">)</span>
<span class="co">#&gt; Task done, CNVR Distribution Plot saved in working directory.</span></code></pre></div>
<div class="figure">
<img src="cnvr_plot_part/cnvr_plot.png" width="320" alt=""><p class="caption">Figure.10 Part CNVR Distribution Map</p>
</div>
</div>
<div id="how-can-we-plot-all-high-frequency-cnvrs-at-once" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-plot-all-high-frequency-cnvrs-at-once" class="anchor"></a>7. How can we plot all high frequency CNVRs at once?</h2>
<p>It always interesting to see which genomic regions have the highest frequencies of structure variation. To view these, we can plot all the high frequency regions using the <code>cnvr_plot</code> function.</p>
<p>As an example, let’s plot the PennCNV results. Remember to set a new folder name to save the results.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cnvr_plot.html">cnvr_plot</a></span><span class="op">(</span>cnvr <span class="op">=</span> <span class="st">"call_cnvr_Penn/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
          clean_cnv <span class="op">=</span> <span class="st">"cnv_clean/penncnv_clean.cnv"</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
          refgene <span class="op">=</span> <span class="st">"refgene/cow_ARS_ENS.txt"</span>, <span class="co">#standard file was generated by 'get_refgene' function in section 5.1</span>
          sample_size <span class="op">=</span> <span class="fl">382</span>, <span class="co">#total sample size</span>
          common_cnv_threshold <span class="op">=</span> <span class="fl">0.18</span>, <span class="co">#the CNV frequency threshold</span>
          width_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#optional, adjust the width of final plot</span>
          height_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#optional, adjust the height of final plot</span>
          col_gene <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional, adjust the colour of annotated genes</span>
          gene_font_size <span class="op">=</span> <span class="fl">2.5</span>, <span class="co">#optional, adjust the font size of annotated gene names</span>
          folder <span class="op">=</span> <span class="st">"high_freq_cnvr_penn"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="high_freq_cnvr_penn/combine_cnvr.png" width="320" alt=""><p class="caption">Figure.11 All high frequency CNVRs from PennCNV</p>
</div>
</div>
<div id="can-we-compare-cnvs-between-different-result-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#can-we-compare-cnvs-between-different-result-sets" class="anchor"></a>8. Can we compare CNVs between different result sets?</h2>
<p>Typically, we will use multiple models and software packages when predicting and analysing CNVs, especially when using data based on SNP chip genotypes. As a result, it is very common that we will need to compare two or more sets of predicted CNVs. To tackle this problem, we have implemented the <code>compare_cnv</code> function to make comparisons on both the individual and population levels. At the individual level, the function checks how many CNVs overlap in the same individual between two CNV result sets, thereby verifying the level of similarity between the two sets of results. On the population level, the comparison will check the level of overlap through the whole sample population, without checking whether or not the CNVs are present in the same individual; this can better indiciate the degree of repeatability of CNV detection between the two result sets.</p>
<p>As an example let’s compare the CNV prediction results for PennCNV and CNVPartition. Note that these results were generated using different reference builds, so the results of the comparison are currently rather meaningless (the physical coordinates of the CNVs will not line up across reference builds): this problem will be fixed in section 11 below.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compare_cnv.html">compare_cnv</a></span><span class="op">(</span>cnv_def <span class="op">=</span> <span class="va">cnv_part_umd</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
            cnv_tar <span class="op">=</span> <span class="va">cnv_penn_ars</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
            width_1 <span class="op">=</span> <span class="fl">12</span>, <span class="co">#optional, adjust the width of final plot</span>
            height_1 <span class="op">=</span> <span class="fl">11</span>, <span class="co">#optional, adjust the height of final plot</span>
            legend_x <span class="op">=</span> <span class="fl">0.9</span>, <span class="co">#optional, adjust the horizontal position of final plot</span>
            legend_y <span class="op">=</span> <span class="fl">0.9</span>, <span class="co">#optional, adjust the vertical position of final plot</span>
            col_1 <span class="op">=</span> <span class="st">"pink"</span>, <span class="co">#optional, adjust the colour of Non-Overlapped part</span>
            col_2 <span class="op">=</span> <span class="st">"lightblue"</span>, <span class="co">#optional, adjust the colour of Overlapped part</span>
            plot_caption <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#optional, show the Note information in plot or not</span>
            folder <span class="op">=</span> <span class="st">"compare_cnv_PartUMD_Vs_PennUMD"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Comparison Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVs which are in total 1239"</span>
<span class="co">#&gt; [1] "The number of overlaped CNVs on individual level is 169, which is around 13.6 percent in the first file"</span>
<span class="co">#&gt; [1] "The number of Non-overlaped CNVs on individual level is 1070, which is around 86.4 percent in the first file"</span>
<span class="co">#&gt; [1] "Comparison to the first file was Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVs which are in total of 1239"</span>
<span class="co">#&gt; [1] "The number of overlaped CNVs on population level is 541, which is around 43.7 percent in first file."</span>
<span class="co">#&gt; [1] "The number of Non-overlaped CNVs on population level is 698, which is around 56.3 percent in first file"</span>
<span class="co">#&gt; [1] "The final comparison results of the first file as follows: "</span>
<span class="co">#&gt;                  Item Individual.Level Population.Level</span>
<span class="co">#&gt; 1     Overlapped CNVs             13.6             43.7</span>
<span class="co">#&gt; 2 Non-overlapped CNVs             86.4             56.3</span>
<span class="co">#&gt; [1] "Comparison to the first file was finished."</span>
<span class="co">#&gt; [1] "Comparison to second file was Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVs which are in total 5215"</span>
<span class="co">#&gt; [1] "The number of overlaped CNVs on individual level is 175, which is around 3.4 percent in second file."</span>
<span class="co">#&gt; [1] "The number of Non-overlaped CNVs on individual level is 5040, which is around 96.6 percent in second file."</span>
<span class="co">#&gt; [1] "Comparison to the second file was Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVs which are in total of 5215"</span>
<span class="co">#&gt; [1] "The number of overlaped CNVs in seceond file on population level are 1379, which is around 26.4 percent"</span>
<span class="co">#&gt; [1] "The number of Non-overlaped CNVs in second file on population level are 3836, which is around 73.6 percent"</span>
<span class="co">#&gt; [1] "The final comparison results of the second file as follows: "</span>
<span class="co">#&gt;                  Item Individual.Level Population.Level</span>
<span class="co">#&gt; 1     Overlapped CNVs              3.4             26.4</span>
<span class="co">#&gt; 2 Non-overlapped CNVs             96.6             73.6</span>
<span class="co">#&gt; [1] "Task done. Comparison results were saved in your working directory"</span></code></pre></div>
<div class="figure">
<img src="compare_cnv_PartUMD_Vs_PennUMD/combine.png" width="330" alt=""><p class="caption">Figure.12 Comparison of PennCNV and CNVPartition results</p>
</div>
</div>
<div id="what-about-cnvrs" class="section level2">
<h2 class="hasAnchor">
<a href="#what-about-cnvrs" class="anchor"></a>9. What about CNVRs?</h2>
<p>Similar to the <code>compare_cnv</code> function, we can use the <code>compare_cnvr</code> function to compare CNV regions. The same caveats apply with regard to the ensuring the two CNVR data sets were built on the same reference genome. Let’s try to use <code>compare_cnvr</code> now with the CNVPartition and Standard data sets (as before, ignoring for now the fact that these were produced on different reference genomes).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compare_cnvr.html">compare_cnvr</a></span><span class="op">(</span>cnvr_def  <span class="op">=</span> <span class="st">"call_cnvr_Part/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
             cnvr_tar  <span class="op">=</span> <span class="st">"call_cnvr_standard/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
             width_1 <span class="op">=</span> <span class="fl">8</span>, <span class="co">#optional, adjust the width of comparison plot</span>
             height_1 <span class="op">=</span> <span class="fl">8</span>, <span class="co">#optional, adjust the height of comparison plot</span>
             hjust_prop <span class="op">=</span> <span class="fl">0.2</span>, <span class="co">#optional, adjust the horizontal position of the proportion number present in plot</span>
             hjust_num <span class="op">=</span> <span class="fl">1.6</span>, <span class="co">#optional, adjust the horizontal position of the number present in plot</span>
             col_1 <span class="op">=</span> <span class="st">"grey15"</span>, <span class="co">#optional, adjust the colour of non-overlapped part</span>
             col_2 <span class="op">=</span> <span class="st">"forestgreen"</span>, <span class="co">#optional, adjust the colour of overlapped part</span>
             folder <span class="op">=</span> <span class="st">"compare_cnvr_PartUMD_Penn_UMD"</span><span class="op">)</span>
<span class="co">#&gt; Comparison to the first file was Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVRs which are in total of 246 </span>
<span class="co">#&gt; CNVR comparison summary results in checkover_pop_1 as following:</span>
<span class="co">#&gt; # A tibble: 6 x 8</span>
<span class="co">#&gt; # Groups:   Type [3]</span>
<span class="co">#&gt;   Type  Check_overlap origi_length num_CNVR overlap_len num_overlap_opp~</span>
<span class="co">#&gt;   &lt;fct&gt; &lt;fct&gt;                &lt;int&gt;    &lt;int&gt;       &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 Gain  Non-Overlap        1824550        7    53469808                0</span>
<span class="co">#&gt; 2 Gain  Overlap          105638930       76    53993672              100</span>
<span class="co">#&gt; 3 Loss  Non-Overlap        6194700       64     6194700                0</span>
<span class="co">#&gt; 4 Loss  Overlap           11580677       64     8208478               64</span>
<span class="co">#&gt; 5 Mixed Non-Overlap        3317779        4     3317779                0</span>
<span class="co">#&gt; 6 Mixed Overlap           70606375       31    37922215               60</span>
<span class="co">#&gt; # ... with 2 more variables: prop_overlap_len &lt;dbl&gt;, prop_num &lt;dbl&gt;</span>
<span class="co">#&gt; The number of overlaped CNVRs on population level is 171, which is around 69.5 percent in first file.</span>
<span class="co">#&gt; The number of Non-overlaped CNVRs on population level is 75, which is around 30.5 percent in first file</span>
<span class="co">#&gt; The overlapping length is 100124365 bp, which is around 50.3 percent in first file</span>
<span class="co">#&gt; The final comparison results of the first file as follows: </span>
<span class="co">#&gt;                   Item Number.of.CNVR Proportion.of.Number.... Length.bp.</span>
<span class="co">#&gt; 1     Overlapped CNVRs            171                     69.5  100124365</span>
<span class="co">#&gt; 2 Non-overlapped CNVRs             75                     30.5   99038646</span>
<span class="co">#&gt; 3             In Total            246                    100.0  199163011</span>
<span class="co">#&gt;   Proportion.of.Length....</span>
<span class="co">#&gt; 1                     50.3</span>
<span class="co">#&gt; 2                     49.7</span>
<span class="co">#&gt; 3                    100.0</span>
<span class="co">#&gt; Comparison to the first file was finished.</span>
<span class="co">#&gt; Comparison to the second file was Passed Validation, the number of Overlap and Non-overlap CNV equal to the original number of CNVRs which are in total of 878</span>
<span class="co">#&gt; CNVR comparison summary results in checkover_pop_2 as following:</span>
<span class="co">#&gt; # A tibble: 6 x 8</span>
<span class="co">#&gt; # Groups:   Type [3]</span>
<span class="co">#&gt;   Type  Check_overlap origi_length num_CNVR overlap_len num_overlap_opp~</span>
<span class="co">#&gt;   &lt;fct&gt; &lt;fct&gt;                &lt;int&gt;    &lt;int&gt;       &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 Gain  Non-Overlap       35340763      238    36131115                0</span>
<span class="co">#&gt; 2 Gain  Overlap           13315879       68    12525527               68</span>
<span class="co">#&gt; 3 Loss  Non-Overlap       31958106      311    31958106                0</span>
<span class="co">#&gt; 4 Loss  Overlap            6279030       51     5426320               51</span>
<span class="co">#&gt; 5 Mixed Non-Overlap       70399922      107    70399922                0</span>
<span class="co">#&gt; 6 Mixed Overlap          129649020      103    82172518              106</span>
<span class="co">#&gt; # ... with 2 more variables: prop_overlap_len &lt;dbl&gt;, prop_num &lt;dbl&gt;</span>
<span class="co">#&gt; The number of overlaped CNVRs in seceond file on population level are 222, which is around 25.3 percent</span>
<span class="co">#&gt; The number of Non-overlaped CNVRs in second file on population level are 656, which is around 74.7 percent</span>
<span class="co">#&gt; The overlapping length is 100124365 bp, which is around 34.9 percent in second file</span>
<span class="co">#&gt; The final comparison results of the second file as follows: </span>
<span class="co">#&gt;                   Item Number.of.CNVR Proportion.of.Number.... Length.bp.</span>
<span class="co">#&gt; 1     Overlapped CNVRs            222                     25.3  100124365</span>
<span class="co">#&gt; 2 Non-overlapped CNVRs            656                     74.7  186818355</span>
<span class="co">#&gt; 3             In Total            878                    100.0  286942720</span>
<span class="co">#&gt;   Proportion.of.Length....</span>
<span class="co">#&gt; 1                     34.9</span>
<span class="co">#&gt; 2                     65.1</span>
<span class="co">#&gt; 3                    100.0</span>
<span class="co">#&gt; Processing the unique common CNVRs list by uniting overlepped CNVRs between two results </span>
<span class="co">#&gt; Chromosome 1 has been processed.</span>
<span class="co">#&gt; Chromosome 2 has been processed.</span>
<span class="co">#&gt; Chromosome 3 has been processed.</span>
<span class="co">#&gt; Chromosome 4 has been processed.</span>
<span class="co">#&gt; Chromosome 5 has been processed.</span>
<span class="co">#&gt; Chromosome 6 has been processed.</span>
<span class="co">#&gt; Chromosome 7 has been processed.</span>
<span class="co">#&gt; Chromosome 8 has been processed.</span>
<span class="co">#&gt; Chromosome 9 has been processed.</span>
<span class="co">#&gt; Chromosome 10 has been processed.</span>
<span class="co">#&gt; Chromosome 11 has been processed.</span>
<span class="co">#&gt; Chromosome 12 has been processed.</span>
<span class="co">#&gt; Chromosome 13 has been processed.</span>
<span class="co">#&gt; Chromosome 14 has been processed.</span>
<span class="co">#&gt; Chromosome 15 has been processed.</span>
<span class="co">#&gt; Chromosome 16 has been processed.</span>
<span class="co">#&gt; Chromosome 17 has been processed.</span>
<span class="co">#&gt; Chromosome 18 has been processed.</span>
<span class="co">#&gt; Chromosome 19 has been processed.</span>
<span class="co">#&gt; Chromosome 20 has been processed.</span>
<span class="co">#&gt; Chromosome 21 has been processed.</span>
<span class="co">#&gt; Chromosome 22 has been processed.</span>
<span class="co">#&gt; Chromosome 23 has been processed.</span>
<span class="co">#&gt; Chromosome 24 has been processed.</span>
<span class="co">#&gt; Chromosome 25 has been processed.</span>
<span class="co">#&gt; Chromosome 26 has been processed.</span>
<span class="co">#&gt; Chromosome 27 has been processed.</span>
<span class="co">#&gt; Chromosome 28 has been processed.</span>
<span class="co">#&gt; Chromosome 29 has been processed.</span>
<span class="co">#&gt; There are 168 unique and mutual CNVRs after uniting the overlapped CNVRs between two results </span>
<span class="co">#&gt; Task done. Comparison results were saved in the working directory</span></code></pre></div>
<div class="figure">
<img src="compare_cnvr_PartUMD_Penn_UMD/combine_cnvr.png" width="320" alt=""><p class="caption">Figure.13 Comparison of CNVR</p>
</div>
</div>
<div id="how-do-we-find-the-consensus-set-of-genes-common-to-multiple-cnv-result-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-find-the-consensus-set-of-genes-common-to-multiple-cnv-result-sets" class="anchor"></a>10. How do we find the consensus set of genes common to multiple CNV result sets?</h2>
<p>One of the most interesting parts of CNV analysis is to find out which genes are affected by structure variations, and to check out the frequency of these aberrant genes in our samples. After using the <code>call_gene</code> function in section 5.2, we now have the gene frequency list for each CNV result set. We can take these lists as input in the <code>compare_gene</code> function to find the set of consensus genes common to multiple CNV results.</p>
<p>Now, let’s compare the annotated gene lists for the PennCNV and CNVPartition results as an example. We can assign the arguments <code>gene_freq_1</code> and <code>gene_freq_2</code> with the <code>gene_freq_cnv.txt</code> files from the <code>call_gene_Part</code> and <code>call_gene_Penn</code> output folders created earlier. Then we need to set an integer threshold (<code>common_gene_threshold</code>) for the number of observations required to declare that a gene is “commom”. In our example, the sample size is 400, so for a prevalence of 5%, the <code>common_gene_threshold</code> could calculated as <code>400 * 5% = 20</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compare_gene.html">compare_gene</a></span><span class="op">(</span>gene_freq_1 <span class="op">=</span> <span class="st">"call_gene_Part/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_2 <span class="op">=</span> <span class="st">"call_gene_Penn_Ens/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             title_1 <span class="op">=</span> <span class="st">"Bovine Part"</span>, <span class="co">#optional, adjust the title of input file 1 present in final plot </span>
             title_2 <span class="op">=</span> <span class="st">"Bovine Penn"</span>, <span class="co">#optional, adjust the title of input file 2 present in final plot </span>
             common_gene_threshold <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional, the threshold is the gene frequency, here 20 means the Gene frequency exceed 20 in two input lists will be regarded as common gene. </span>
             col_1 <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, set the colour of 'Common High' genes</span>
             col_2 <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional, set the colour of 'Common Low' genes</span>
             col_3 <span class="op">=</span> <span class="st">"gold"</span>, <span class="co">#optional, set the colour of 'gene_freq_1'-only genes</span>
             col_4 <span class="op">=</span> <span class="st">"green"</span>, <span class="co">#optional, set the colour of 'gene_freq_2'-only genes</span>
             height_1 <span class="op">=</span> <span class="fl">10</span>, <span class="co">#optional, adjust the height of final plot</span>
             width_1 <span class="op">=</span>  <span class="fl">12</span>, <span class="co">#optional, adjust the width of final </span>
             folder <span class="op">=</span> <span class="st">"compare_gene_Part_Vs_Penn"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Clustering common gene between the two input files..."</span>
<span class="co">#&gt; [1] "Brief summary of comparison as below:"</span>
<span class="co">#&gt; # A tibble: 4 x 2</span>
<span class="co">#&gt; # Groups:   Common_Gene [4]</span>
<span class="co">#&gt;   Common_Gene          n</span>
<span class="co">#&gt;   &lt;chr&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 Common_High          5</span>
<span class="co">#&gt; 2 Common_Low        3029</span>
<span class="co">#&gt; 3 High_Freq_list_1     6</span>
<span class="co">#&gt; 4 High_Freq_list_2    54</span>
<span class="co">#&gt; [1] "Gene comparison list, brife summary and comparison plot was saved in working directory."</span></code></pre></div>
<div class="figure">
<img src="compare_gene_Part_Vs_Penn/two_lists_comparison.png" width="300" alt=""><p class="caption">Figure.14 Comparison of genes present in the PennCNV (green) and CNVPartition (gold) gene lists</p>
</div>
<p>We can also compare the frequency lists for three genes, just by adding one more input file (using the <code>gene_freq_3</code> argument); in this case, the comparison figure will be a 3D plot.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compare_gene.html">compare_gene</a></span><span class="op">(</span>gene_freq_1 <span class="op">=</span> <span class="st">"call_gene_Part/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_2 <span class="op">=</span> <span class="st">"call_gene_Penn_Ens/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_3 <span class="op">=</span> <span class="st">"call_gene_standard_ARS_UCSC/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             title_1 <span class="op">=</span> <span class="st">"Part"</span>, <span class="co">#optional, adjust the title of input file 1 present in final plot </span>
             title_2 <span class="op">=</span> <span class="st">"ENS"</span>, <span class="co">#optional, adjust the title of input file 2 present in final plot </span>
             title_3 <span class="op">=</span> <span class="st">"UCSC"</span>, <span class="co">#optional, adjust the title of input file 3 present in final plot </span>
             common_gene_threshold <span class="op">=</span> <span class="fl">15</span>,
             height_1 <span class="op">=</span> <span class="fl">5.8</span>, <span class="co">#optional, adjust the height of final plot</span>
             width_1 <span class="op">=</span> <span class="fl">5.8</span>, <span class="co">#optional, adjust the width of final plot</span>
             folder <span class="op">=</span> <span class="st">"compare_gene_three_lists"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Starting merging gene lists..."</span>
<span class="co">#&gt; [1] "Clustering common gene among three gene lists..."</span>
<span class="co">#&gt; [1] "2 genes with higher frequency in all three gene lists, they are: "</span>
<span class="co">#&gt;      name2_1 Frequency_1 Frequency_2 Frequency_3 Common_Gene</span>
<span class="co">#&gt; 1211   CERS6          21          24          23 Common_High</span>
<span class="co">#&gt; 3354   STK39          21          17          19 Common_High</span>
<span class="co">#&gt; [1] "Brief summary:"</span>
<span class="co">#&gt; # A tibble: 3 x 2</span>
<span class="co">#&gt; # Groups:   Common_Gene [3]</span>
<span class="co">#&gt;   Common_Gene  Number_Gene</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;</span>
<span class="co">#&gt; 1 Common_High            2</span>
<span class="co">#&gt; 2 Common_Low          3783</span>
<span class="co">#&gt; 3 High_and_Low         115</span>
<span class="co">#&gt; [1] "Gene comparison list, brife summary and comparison plot was saved in working directory."</span></code></pre></div>
<div class="figure">
<img src="compare_gene_three_lists/threegene.png" width="320" alt=""><p class="caption">Figure.15 Comparison of three gene lists</p>
</div>
<p>If we add a fourth gene frequency lists, by adding the <code>gene_freq_4</code> argument, the comparison figure produced will be a heat-map rather than a scatterplot.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compare_gene.html">compare_gene</a></span><span class="op">(</span>gene_freq_1 <span class="op">=</span> <span class="st">"call_gene_Part/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_2 <span class="op">=</span> <span class="st">"call_gene_Penn_UCSC/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_3 <span class="op">=</span> <span class="st">"call_gene_Part/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             gene_freq_4 <span class="op">=</span> <span class="st">"call_gene_standard_ARS_UCSC/gene_freq_cnv.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
             title_1 <span class="op">=</span> <span class="st">"Part"</span>, <span class="co">#optional, adjust the title of input file 1 present in final plot </span>
             title_2 <span class="op">=</span> <span class="st">"ENS"</span>, <span class="co">#optional, adjust the title of input file 2 present in final plot </span>
             title_3 <span class="op">=</span> <span class="st">"UCSC"</span>, <span class="co">#optional, adjust the title of input file 3 present in final plot </span>
             title_4 <span class="op">=</span> <span class="st">"PUCSC"</span>, <span class="co">#optional, adjust the title of input file 4 present in final plot </span>
             col_1 <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, set the color of High frequency genes</span>
             col_2 <span class="op">=</span> <span class="st">"yellow"</span>, <span class="co">#optional, set the color of Low frequency genes</span>
             common_gene_threshold <span class="op">=</span> <span class="fl">20</span>,
             height_1 <span class="op">=</span> <span class="fl">9</span>, <span class="co">#optional, adjust the height of final plot</span>
             width_1 <span class="op">=</span> <span class="fl">11</span>, <span class="co">#optional, adjust the width of final plot</span>
             folder <span class="op">=</span> <span class="st">"compare_gene_four_lists"</span><span class="op">)</span>
<span class="co">#&gt; Starting merging gene lists...</span>
<span class="co">#&gt; Clustering common gene among four gene lists...</span>
<span class="co">#&gt; 1 genes with higher frequency in all gene lists, they are: </span>
<span class="co">#&gt;     name2_1 Frequency_1 Frequency_2 Frequency_3 Frequency_4 Common_check</span>
<span class="co">#&gt; 329   CERS6          21          24          21          23  Common_high</span>
<span class="co">#&gt; Brief summary:</span>
<span class="co">#&gt; # A tibble: 3 x 2</span>
<span class="co">#&gt; # Groups:   Common_check [3]</span>
<span class="co">#&gt;   Common_check Number_Gene</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;</span>
<span class="co">#&gt; 1 Common_high            1</span>
<span class="co">#&gt; 2 Common_low          2401</span>
<span class="co">#&gt; 3 High_and_Low          34</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Gene comparison list, brife summary and comparison plot was saved in working directory.</span></code></pre></div>
<div class="figure">
<img src="compare_gene_four_lists/four_gene_heatmap.png" width="300" alt=""><p class="caption">Figure.16 Heatmap comparing gene lists across four data sets</p>
</div>
</div>
<div id="brief-summary" class="section level2">
<h2 class="hasAnchor">
<a href="#brief-summary" class="anchor"></a>Brief summary</h2>
<p>The ten functions we’ve run so far only require users to provide the CNV results, which can already generate plenty of summary results that include sufficient information to point us in the right direction to dig deeper. If we want to explore a particular region in more depth, there are extra functions in the <code>HandyCNV</code> package that can combine CNV data with other data sources. We will introduce some of these functions in the following section.</p>
</div>
<div id="how-can-we-create-the-map-file-used-in-handycnv-to-allow-comparing-cnvs-between-different-reference-genomes" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-create-the-map-file-used-in-handycnv-to-allow-comparing-cnvs-between-different-reference-genomes" class="anchor"></a>11. How can we create the map file used in <code>HandyCNV</code>, to allow comparing CNVs between different reference genomes?</h2>
<p>We can use the <code>convert_map</code> function to prepare a map file; generally, we will need one for each reference file. In our examples, we have used the two bovine reference genomes ARS-UCD 1.2 and UMD 3.1. For <code>HandyCNV</code>, map files hould have the format shown in the table below, with the four columns <code>Chromosome</code>, <code>SNP</code>, <code>MorganPos</code>， and <code>PhysicalPos</code>. If the centiMorgan position (<code>MorganPos</code>) is unknown, we can assign the missing value as <code>0</code>, or provide the phyisical position in megabases. This format is the same as the standard <code>.map</code> format used in Plink and other software, and can be exported from GenomeStudio. The column names are not required.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tabl</span> <span class="op">&lt;-</span> <span class="st">"
-----------------------------------------
Chr   SNP           MorganPos PhysicalPos
---- ------------- ---------- -----------               
  5   14636         56.3275   56327514 
  5   16084         56.329    56328962  
  5   19597         56.3325   56332475  
  1   1_115292065   114.371   114370731  
  1   1_115292107   114.371   114370773 
-----------------------------------------"</span></code></pre></div>
<p>Here, we have the map files for both bovine references included in the package, at the following paths:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ars_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/map/ARS1.2_GGPHDV3.map"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span>
<span class="va">umd_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/map/UMD3.1_Bovine.map"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span></code></pre></div>
<p>The function will not only generate map files, for use in <code>compare_cnv</code> and <code>compare_cnvr</code>, it will also generate map file used in Plink and PennCNV for both inputs. The conversion process also compares the two input maps and produces summary tables and figures.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/convert_map.html">convert_map</a></span><span class="op">(</span>default_map <span class="op">=</span> <span class="va">umd_map</span>, <span class="co">#when running your own data, set this argument to 'default_map = "localpath/file"'</span>
            target_map <span class="op">=</span> <span class="va">ars_map</span>, <span class="co">#when running your own data, set this argument to 'target_map = "localpath/file"'</span>
            chr_set <span class="op">=</span> <span class="fl">30</span>,
            defMap_title <span class="op">=</span> <span class="st">"UMD 3.1"</span>, <span class="co">#set the label of axis</span>
            tarMap_title <span class="op">=</span> <span class="st">"ARS-UCD 1.2"</span>, <span class="co">#set the label of axis</span>
            col_1 <span class="op">=</span> <span class="st">"green4"</span>, <span class="co">#set colour for the type of Match in SNP comparison plot</span>
            col_2 <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#set colour for the type of Unmatched in SNP comparison plot</span>
            col_3 <span class="op">=</span> <span class="st">"deeppink2"</span>, <span class="co">#set colour for the bar of Target_Map in SNP density plot</span>
            col_4 <span class="op">=</span> <span class="st">"deeppink2"</span>, <span class="co">#set colour for line of Target_map in density plot</span>
            col_5 <span class="op">=</span> <span class="st">"turquoise3"</span>, <span class="co">#set colour for point of Default_Map in SNP density plot</span>
            col_6 <span class="op">=</span> <span class="st">"turquoise3"</span>, <span class="co">#set colour for line of Default_map in SNP density plot</span>
            species <span class="op">=</span> <span class="st">"Bovine"</span>, 
            height_c <span class="op">=</span> <span class="fl">16</span>,
            width_c <span class="op">=</span> <span class="fl">22</span>,
            height_d <span class="op">=</span> <span class="fl">14</span>,
            width_d <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>
<span class="co">#&gt; [1] "Starting comparing the differences between the two versions"</span>
<span class="co">#&gt; [1] "Checking if negtive position exsits in new converted map...."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  FALSE   TRUE </span>
<span class="co">#&gt; 139512    465 </span>
<span class="co">#&gt; [1] "There are 465 SNPs with negtive postion in new assembly."</span>
<span class="co">#&gt; [1] "The matching results of SNP chromosome ID in the two versions are as follows:"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Match Mismatch </span>
<span class="co">#&gt;   122772    17211 </span>
<span class="co">#&gt; [1] "The details of comparison as follows: "</span>
<span class="co">#&gt;                           Type.Var1 Type.Freq</span>
<span class="co">#&gt; 1           Both unknown Chromosome       344</span>
<span class="co">#&gt; 2              Different Chromosome       851</span>
<span class="co">#&gt; 3                   Same Chromosome    122428</span>
<span class="co">#&gt; 4            Unfound in Default map      2038</span>
<span class="co">#&gt; 5             Unfound in Target map         6</span>
<span class="co">#&gt; 6 Unknown Chromosome in Default map     13771</span>
<span class="co">#&gt; 7  Unknown Chromosome in Target map       545</span>
<span class="co">#&gt; [1] "Ploting SNP comparison results....."</span>
<span class="co">#&gt; [1] "Comparison of SNP by Chromosome plot was saved in working directory."</span>
<span class="co">#&gt; [1] "Preparing the map files for PennCNV and Plink...."</span>
<span class="co">#&gt; [1] "Writing converted map for PennCNV and Plink formats..."</span>
<span class="co">#&gt; [1] "Target map was converted to PennCNV map file..."</span>
<span class="co">#&gt; [1] "Converted map for PennCNV format were saved in working directory."</span>
<span class="co">#&gt; [1] "Converted map for Plink format were saved in working directory."</span>
<span class="co">#&gt; [1] "Plotting the Difference of SNPs between two versions of map..."</span>
<span class="co">#&gt; [1] "SNP difference plot was saved in working directory."</span></code></pre></div>
<div class="figure">
<img src="convert_map/comparison_snp.png" width="320" alt=""><p class="caption">Figure.17 Comparison of the ARS 1.2 and UMD 3.1 reference genomes</p>
</div>
<div class="figure">
<img src="convert_map/SNP_difference_density.png" width="320" alt=""><p class="caption">Figure.18 Comparison of SNP densities on the ARS 1.2 and UMD 3.1 maps</p>
</div>
</div>
<div id="how-can-we-integrate-the-cnvs-and-annotated-genes-with-additional-information-such-as-log-r-ratio-b-allele-frequency-call-rate-heterozygosity-missing-value-rate-and-linkage-disequilibrium-and-plot-it-as-one-figure" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-integrate-the-cnvs-and-annotated-genes-with-additional-information-such-as-log-r-ratio-b-allele-frequency-call-rate-heterozygosity-missing-value-rate-and-linkage-disequilibrium-and-plot-it-as-one-figure" class="anchor"></a>12. How can we integrate the CNVs and annotated genes with additional information, such as Log R ratio, B Allele Frequency, call rate, heterozygosity, missing value rate and Linkage Disequilibrium, and plot it as one figure?</h2>
<p>The <code>plot_cnvr_panorama</code> function was built to achieve this task. But we need provide more information to this function: a. CNVR results(generated by ‘call_cnvr’ in section 4) b. Annotated CNV file (generated by ‘call_gene’ function) c. Signal Intensity file (In general it export from GenomeStudio which used in CNV detection process) d. Map file (Four columns used in Plink or generated by ‘convert_map’ function) e. Genotype files (generated by Plink with ‘–make-bed’ flag which include BED, BIM, FAM three files)</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cnvr_panorama.html">plot_cnvr_panorama</a></span><span class="op">(</span>cnvr <span class="op">=</span> <span class="st">"call_cnvr_Penn/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
                   cnv_annotation <span class="op">=</span> <span class="st">"call_gene_Penn_UCSC/cnv_annotation.txt"</span>, <span class="co">#standard file was generated by 'call_gene' function in Section 5.2</span>
                   intensity <span class="op">=</span> <span class="st">"Demo_data/demo_intensity.txt"</span>, 
                   map <span class="op">=</span> <span class="st">"Demo_data/demo.map"</span>, 
                   prefix_bed <span class="op">=</span> <span class="st">"Demo_data/demo_HandyCNV"</span>, 
                   ld_heat <span class="op">=</span> <span class="cn">TRUE</span>,
                   sample_size <span class="op">=</span> <span class="fl">382</span>, 
                   common_cnv_threshold <span class="op">=</span> <span class="fl">0.18</span>, 
                   col_0 <span class="op">=</span> <span class="st">"green"</span>, <span class="co">#optional, customize color for o copy CNV </span>
                   col_1 <span class="op">=</span> <span class="st">"hotpink"</span>, <span class="co">#optional, customize color for 1 copy CNV </span>
                   col_2 <span class="op">=</span> <span class="st">"NA"</span>, <span class="co">#optional, customize color for 2 copy CNV</span>
                   col_3 <span class="op">=</span> <span class="st">"turquoise"</span>, <span class="co">#optional, customize color for 3 copy CNV </span>
                   col_4 <span class="op">=</span> <span class="st">"aquamarine"</span>, <span class="co">#optional, customize color for 4 copy CNV </span>
                   col_gene <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional, customize color of gene </span>
                   width_1 <span class="op">=</span> <span class="fl">16</span>,
                   height_1 <span class="op">=</span> <span class="fl">30</span>,
                   folder <span class="op">=</span> <span class="st">"cnvr_panorama_heat_ld"</span><span class="op">)</span>
<span class="co">#&gt; [1] "4 CNVRs passed the common frequency threshold."</span>
<span class="co">#&gt; Reading Demo_data/demo_HandyCNV.fam </span>
<span class="co">#&gt; Reading Demo_data/demo_HandyCNV.bim </span>
<span class="co">#&gt; Reading Demo_data/demo_HandyCNV.bed </span>
<span class="co">#&gt; ped stats and snps stats have been set. </span>
<span class="co">#&gt; 'p' has been set. </span>
<span class="co">#&gt; 'mu' and 'sigma' have been set.</span>
<span class="co">#&gt; [1] "Starting to make plot..."</span>
<span class="co">#&gt; [1] "Plot 1 (7_310937_435584_Mixed.png) was stored in the cnvr_panorama_heat_ld/ directory."</span>
<span class="co">#&gt; [1] "Plot 2 (15_78804694_79611716_Mixed.png) was stored in the cnvr_panorama_heat_ld/ directory."</span>
<span class="co">#&gt; [1] "Plot 3 (12_69804346_70777223_Mixed.png) was stored in the cnvr_panorama_heat_ld/ directory."</span>
<span class="co">#&gt; [1] "Plot 4 (12_56513728_64796304_Mixed.png) was stored in the cnvr_panorama_heat_ld/ directory."</span></code></pre></div>
<div class="figure">
<img src="cnvr_panorama_heat_ld/7_310937_435584_Mixed.png" width="150" alt=""><p class="caption">Figure.19 Example of CNVR Panorama with LD_Heatmap</p>
</div>
<p>For LD, we can replace the heatmap with the classic inverted triangle diagram by setting the argument <code>ld_heat = FALSE</code>, which will use the <code>gaston</code> package to plot LD. Because this requires converting the figure to <code>ggplot</code> format (via the <code>ggplotify</code> package), this process can be time consuming, and sometimes does not work well for large genomic regions. Therefore, we suggest that you use <code>ld_heat = TRUE</code> initially, before narrowing in on a particular region of interest to display with the classic LD plot.</p>
<div class="figure">
<img src="cnvr_panorama_classical_ld/7_310937_435584_Mixed.png" width="150" alt=""><p class="caption">Figure.20 Example of CNVR Panorama with Classical_LD</p>
</div>
</div>
<div id="how-can-we-make-a-plot-to-show-the-source-of-the-cnvs" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-make-a-plot-to-show-the-source-of-the-cnvs" class="anchor"></a>13. How can we make a plot to show the source of the CNVs?</h2>
<p>CNVs are heritable: with the pedigree information, particularly for the sires in a livestock context, we can produce plots of CNV sources. The <code>plot_cnvr_source</code> function can produce plots by sire, herd, and sire source. Sire source can be any freeform text, for example the country of origin. The expected input format is comma-separated (with a header), containing the following columns (in any order):</p>
<ul>
<li>
<code>Chip_ID</code> (matched to the <code>Sample_ID</code> column in the CNVR data)</li>
<li><code>Herd</code></li>
<li><code>Sire_ID</code></li>
<li><code>Sire_Source</code></li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cnvr_source.html">plot_cnvr_source</a></span><span class="op">(</span>cnvr <span class="op">=</span> <span class="st">"call_cnvr_Penn/cnvr.txt"</span>, <span class="co">#standard file was generated by 'call_cnvr' function in section 4</span>
                clean_cnv <span class="op">=</span> <span class="st">"cnv_clean/penncnv_clean.cnv"</span>, <span class="co">#standard file was generated by 'cnv_clean' function in section 1</span>
                pedigree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Demo_data/Pedigree/Pedigree.csv"</span>, package <span class="op">=</span> <span class="st">"HandyCNV"</span><span class="op">)</span>, 
                Frequent_threshold <span class="op">=</span> <span class="fl">80</span>,
                col_0 <span class="op">=</span> <span class="st">"springgreen"</span>, col_1 <span class="op">=</span> <span class="st">"pink"</span>, col_3 <span class="op">=</span> <span class="st">"lightblue"</span>, col_4 <span class="op">=</span> <span class="st">"red"</span>,
                report_id <span class="op">=</span> <span class="cn">T</span>,
                height_1 <span class="op">=</span> <span class="fl">20</span>, 
                width_1 <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="plot_cnvr_source/Chr12_56.513728-64.796304Mb_Source.png" width="150" alt=""><p class="caption">Figure.21 Plot Source of CNVR</p>
</div>
</div>
<div id="how-can-we-plot-just-the-genes-in-a-specific-region-to-save-as-a-seperate-figure" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-plot-just-the-genes-in-a-specific-region-to-save-as-a-seperate-figure" class="anchor"></a>14. How can we plot just the genes in a specific region, to save as a seperate figure?</h2>
<p>The <code>plot_gene</code> function can do this for a specific reference gene set and genomic region.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_gene.html">plot_gene</a></span><span class="op">(</span>refgene <span class="op">=</span> <span class="st">"refgene/cow_ARS_UCSC.txt"</span>, 
          chr_id <span class="op">=</span> <span class="fl">7</span>, 
          start <span class="op">=</span> <span class="fl">11</span>, 
          end <span class="op">=</span> <span class="fl">12</span>, 
          height_1 <span class="op">=</span> <span class="fl">3</span>, 
          width_1 <span class="op">=</span> <span class="fl">10</span>, 
          show_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">11.0</span>, <span class="fl">11.35</span>, <span class="fl">11.75</span>, <span class="fl">12.00</span><span class="op">)</span>, 
          gene_font_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Chr7_11_12.png" width="300" alt=""><p class="caption">Figure.22 Gene plot</p>
</div>
</div>
<div id="how-can-we-find-regions-with-high-frequencies-of-runs-of-homozygosity" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-we-find-regions-with-high-frequencies-of-runs-of-homozygosity" class="anchor"></a>15. How can we find regions with high frequencies of runs of homozygosity?</h2>
<p>An ROH (Run of homozygosity) is a region of an individual’s genome where almost all variants have homozygous genotypes. These regions generally indicate the presence of inbreeding in that individual’s ancestry, with more recent inbreeding leading to longer ROHs.</p>
<p>The function <code>roh_window</code> will produce two summary tables, as well as plotting the high frequency regions by chromosome. The first table summarises the total numbers and lengths of ROHs for each individual; the second reports summary statistics for aggregated windows of ROHs across all individuals.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_window.html">roh_window</a></span><span class="op">(</span>roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           window_size <span class="op">=</span> <span class="fl">2</span>,
           length_autosomal <span class="op">=</span> <span class="fl">2489.386</span>, 
           threshold <span class="op">=</span> <span class="fl">0.15</span>, 
           folder <span class="op">=</span> <span class="st">"roh_window"</span>, 
           col_higher <span class="op">=</span> <span class="st">"red"</span>, 
           col_lower <span class="op">=</span> <span class="st">"black"</span>, 
           height_1 <span class="op">=</span> <span class="fl">15</span>, 
           width_1 <span class="op">=</span> <span class="fl">11</span>, 
           legend_x <span class="op">=</span> <span class="fl">0.92</span>, 
           legend_y <span class="op">=</span> <span class="fl">0.06</span>,
           ncol_1 <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt; To group by length...</span>
<span class="co">#&gt; The summary of length groups as following: </span>
<span class="co">#&gt; # A tibble: 5 x 8</span>
<span class="co">#&gt;   Length_group number_ROH    Mean      SD     Min    Max Proportion N_per_animal</span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1 &gt;=16Mb              635  2.22e7 6454382  1.60e7 5.51e7      0.036          1.6</span>
<span class="co">#&gt; 2 1-2Mb              4193  1.58e6  255957  1.00e6 2.00e6      0.239         10.6</span>
<span class="co">#&gt; 3 2-4Mb              6239  2.86e6  574522  2.00e6 4.00e6      0.356         15.7</span>
<span class="co">#&gt; 4 4-8Mb              4297  5.62e6 1139048  4.00e6 8.00e6      0.245         10.8</span>
<span class="co">#&gt; 5 8-16Mb             2151  1.09e7 2202856  8.00e6 1.60e7      0.123          5.4</span>
<span class="co">#&gt; Calculating inbreeding coefficient...</span>
<span class="co">#&gt; by length group...</span>
<span class="co">#&gt; Inbreeding coefficient of Individual as following:</span>
<span class="co">#&gt; # A tibble: 397 x 12</span>
<span class="co">#&gt;    Sample_ID     n total_length  Mean    SD min_length max_length Froh_1Mb</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;        &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1 20109456~    56        285.   5.08  4.2        1.19      21.8     0.114</span>
<span class="co">#&gt;  2 20109456~    37        176.   4.77  3.63       1.24      17.4     0.071</span>
<span class="co">#&gt;  3 20109456~    59        258.   4.37  3.46       1.11      14.4     0.104</span>
<span class="co">#&gt;  4 20109456~    35        170.   4.86  3.78       1.06      15.5     0.068</span>
<span class="co">#&gt;  5 20109456~    71        324.   4.56  3.99       1.11      22.2     0.13 </span>
<span class="co">#&gt;  6 20109456~    61        279.   4.58  3.38       1.03      16.0     0.112</span>
<span class="co">#&gt;  7 20109456~    91        573.   6.29  5.44       1.06      26.1     0.23 </span>
<span class="co">#&gt;  8 20109456~    54        376.   6.96  5.83       1.09      32.1     0.151</span>
<span class="co">#&gt;  9 20109456~    19         51.1  2.69  1.88       1.36       9.67    0.021</span>
<span class="co">#&gt; 10 20109456~    61        336.   5.5   5.68       1.19      30.7     0.135</span>
<span class="co">#&gt; # ... with 387 more rows, and 4 more variables: Froh_2Mb &lt;dbl&gt;, Froh_4Mb &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Froh_8Mb &lt;dbl&gt;, Froh_16Mb &lt;dbl&gt;</span>
<span class="co">#&gt; Setting the sliding window in 2 Mb interval...</span>
<span class="co">#&gt; Counting the frequency of ROHs that overlapped to sliding windows...</span>
<span class="co">#&gt; The top 10 ROHs by frequency are as follows:</span>
<span class="co">#&gt;    window_id number_roh Chr Start_win End_win prop_roh</span>
<span class="co">#&gt; 1    Win_355        220   6   3.8e+07 4.0e+07    0.554</span>
<span class="co">#&gt; 2    Win_649        170  11   6.6e+07 6.8e+07    0.428</span>
<span class="co">#&gt; 3    Win_354        150   6   3.6e+07 3.8e+07    0.378</span>
<span class="co">#&gt; 4    Win_382        143   6   9.2e+07 9.4e+07    0.360</span>
<span class="co">#&gt; 5    Win_381        142   6   9.0e+07 9.2e+07    0.358</span>
<span class="co">#&gt; 6    Win_366        135   6   6.0e+07 6.2e+07    0.340</span>
<span class="co">#&gt; 7    Win_380        133   6   8.8e+07 9.0e+07    0.335</span>
<span class="co">#&gt; 8    Win_650        132  11   6.8e+07 7.0e+07    0.332</span>
<span class="co">#&gt; 9    Win_648        131  11   6.4e+07 6.6e+07    0.330</span>
<span class="co">#&gt; 10   Win_859        123  16   2.8e+07 3.0e+07    0.310</span>
<span class="co">#&gt; Uniting the adjacent sliding windows that with ROH frequency greater than  15 % of the sample size...</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 1 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 2 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 3 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 4 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 5 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 6 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 7 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 8 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 9 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 11 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 12 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 13 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 14 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 15 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 16 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 17 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 18 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 19 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 20 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 21 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 22 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 23 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 24 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 25 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 27 has been processed.</span>
<span class="co">#&gt; High frequency ROH regions on Chromosome 29 has been processed.</span>
<span class="co">#&gt;    Interval_ID Chr    Start      End   Length</span>
<span class="co">#&gt; 1        ROH_1   1 6.40e+07 6.60e+07  2000001</span>
<span class="co">#&gt; 2        ROH_2   1 7.60e+07 8.00e+07  4000001</span>
<span class="co">#&gt; 3        ROH_3   1 1.54e+08 1.56e+08  2000001</span>
<span class="co">#&gt; 4        ROH_4   2 8.00e+07 8.20e+07  2000001</span>
<span class="co">#&gt; 5        ROH_5   3 8.40e+07 9.00e+07  6000001</span>
<span class="co">#&gt; 6        ROH_6   4 2.00e+06 1.20e+07 10000001</span>
<span class="co">#&gt; 7        ROH_7   4 1.40e+07 1.60e+07  2000001</span>
<span class="co">#&gt; 8        ROH_8   4 1.80e+07 2.00e+07  2000001</span>
<span class="co">#&gt; 9        ROH_9   4 3.00e+07 3.80e+07  8000001</span>
<span class="co">#&gt; 10      ROH_10   4 9.60e+07 1.00e+08  4000001</span>
<span class="co">#&gt; 11      ROH_11   5 1.00e+07 2.60e+07 16000001</span>
<span class="co">#&gt; 12      ROH_12   5 3.20e+07 3.40e+07  2000001</span>
<span class="co">#&gt; 13      ROH_13   5 3.60e+07 4.00e+07  4000001</span>
<span class="co">#&gt; 14      ROH_14   5 4.20e+07 5.00e+07  8000001</span>
<span class="co">#&gt; 15      ROH_15   5 6.00e+07 8.80e+07 28000001</span>
<span class="co">#&gt; 16      ROH_16   5 9.60e+07 1.00e+08  4000001</span>
<span class="co">#&gt; 17      ROH_17   6 3.20e+07 1.06e+08 74000001</span>
<span class="co">#&gt; 18      ROH_18   6 1.12e+08 1.16e+08  4000001</span>
<span class="co">#&gt; 19      ROH_19   7 4.00e+07 4.40e+07  4000001</span>
<span class="co">#&gt; 20      ROH_20   8 1.06e+08 1.08e+08  2000001</span>
<span class="co">#&gt; 21      ROH_21   9 0.00e+00 2.00e+06  2000001</span>
<span class="co">#&gt; 22      ROH_22   9 2.00e+07 2.40e+07  4000001</span>
<span class="co">#&gt; 23      ROH_23  11 0.00e+00 4.00e+06  4000001</span>
<span class="co">#&gt; 24      ROH_24  11 1.00e+07 2.60e+07 16000001</span>
<span class="co">#&gt; 25      ROH_25  11 2.80e+07 4.20e+07 14000001</span>
<span class="co">#&gt; 26      ROH_26  11 5.40e+07 6.00e+07  6000001</span>
<span class="co">#&gt; 27      ROH_27  11 6.20e+07 7.20e+07 10000001</span>
<span class="co">#&gt; 28      ROH_28  11 8.60e+07 9.20e+07  6000001</span>
<span class="co">#&gt; 29      ROH_29  11 1.02e+08 1.04e+08  2000001</span>
<span class="co">#&gt; 30      ROH_30  12 2.00e+06 4.00e+06  2000001</span>
<span class="co">#&gt; 31      ROH_31  12 8.80e+07 9.00e+07  2000001</span>
<span class="co">#&gt; 32      ROH_32  13 3.40e+07 4.00e+07  6000001</span>
<span class="co">#&gt; 33      ROH_33  13 4.20e+07 6.00e+07 18000001</span>
<span class="co">#&gt; 34      ROH_34  14 5.00e+07 5.80e+07  8000001</span>
<span class="co">#&gt; 35      ROH_35  14 6.40e+07 6.60e+07  2000001</span>
<span class="co">#&gt; 36      ROH_36  14 7.80e+07 8.20e+07  4000001</span>
<span class="co">#&gt; 37      ROH_37  15 3.00e+07 3.20e+07  2000001</span>
<span class="co">#&gt; 38      ROH_38  15 3.40e+07 3.60e+07  2000001</span>
<span class="co">#&gt; 39      ROH_39  16 1.60e+07 3.80e+07 22000001</span>
<span class="co">#&gt; 40      ROH_40  16 4.00e+07 4.60e+07  6000001</span>
<span class="co">#&gt; 41      ROH_41  16 4.80e+07 5.00e+07  2000001</span>
<span class="co">#&gt; 42      ROH_42  17 4.00e+07 4.80e+07  8000001</span>
<span class="co">#&gt; 43      ROH_43  17 6.00e+07 6.20e+07  2000001</span>
<span class="co">#&gt; 44      ROH_44  18 3.60e+07 4.20e+07  6000001</span>
<span class="co">#&gt; 45      ROH_45  19 2.80e+07 4.00e+07 12000001</span>
<span class="co">#&gt; 46      ROH_46  19 4.40e+07 5.60e+07 12000001</span>
<span class="co">#&gt; 47      ROH_47  20 2.20e+07 2.80e+07  6000001</span>
<span class="co">#&gt; 48      ROH_48  21 8.00e+06 1.00e+07  2000001</span>
<span class="co">#&gt; 49      ROH_49  21 2.60e+07 2.80e+07  2000001</span>
<span class="co">#&gt; 50      ROH_50  21 3.20e+07 3.60e+07  4000001</span>
<span class="co">#&gt; 51      ROH_51  21 4.40e+07 5.00e+07  6000001</span>
<span class="co">#&gt; 52      ROH_52  21 6.80e+07 7.00e+07  2000001</span>
<span class="co">#&gt; 53      ROH_53  22 4.00e+06 1.00e+07  6000001</span>
<span class="co">#&gt; 54      ROH_54  23 0.00e+00 2.00e+06  2000001</span>
<span class="co">#&gt; 55      ROH_55  24 3.40e+07 3.60e+07  2000001</span>
<span class="co">#&gt; 56      ROH_56  24 6.00e+07 6.20e+07  2000001</span>
<span class="co">#&gt; 57      ROH_57  25 0.00e+00 4.00e+06  4000001</span>
<span class="co">#&gt; 58      ROH_58  27 2.00e+06 4.00e+06  2000001</span>
<span class="co">#&gt; 59      ROH_59  29 2.60e+07 3.00e+07  4000001</span>
<span class="co">#&gt; 60      ROH_60  29 3.20e+07 3.40e+07  2000001</span>
<span class="co">#&gt; The total length of high frequency ROH regions about 414000060 bp.</span>
<span class="co">#&gt; Plotting frequncy of ROH by windows...</span>
<span class="co">#&gt; Task done.</span></code></pre></div>
<div class="figure">
<img src="roh_window/roh_freq_windows.png" width="300" alt=""><p class="caption">Figure.23 ROH Frequency Distribution</p>
</div>
</div>
<div id="how-do-we-visualise-rohs" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-visualise-rohs" class="anchor"></a>16. How do we visualise ROHs?</h2>
<p>Visualising ROH is similar to visualising CNVs. The exception is that, because length is particularly informative for ROHs (about the age of inbreeding), the ROHs are coloured by their lengths. The <code>roh_visual</code> function can plot ROHs, either across the whole genome, or chromosome by chromosome. It also supports restricting plotting to regions of interest, and allows adding a panel of gene locations.</p>
<div id="plot-roh-by-chromsome" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-roh-by-chromsome" class="anchor"></a>16.1 Plot ROH by Chromsome</h3>
<p>To plot a single chromosome, use the <code>chr_id</code> argument.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           chr_id <span class="op">=</span> <span class="fl">6</span>, 
           width_1 <span class="op">=</span> <span class="fl">14</span>, 
           height_1 <span class="op">=</span> <span class="fl">9</span>, 
           folder <span class="op">=</span> <span class="st">"roh_visual"</span>, 
           col_short <span class="op">=</span> <span class="st">"red"</span>, 
           col_mid <span class="op">=</span> <span class="st">"blue"</span>, 
           col_long <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>
<span class="co">#&gt; Preparing plot data...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="roh_visual/Chr11_roh.png" width="300" alt=""><p class="caption">Figure.24 ROH on Chromosome 6</p>
</div>
</div>
<div id="plot-roh-by-single-region" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-roh-by-single-region" class="anchor"></a>16.2 Plot ROH by single region</h3>
<p>To plot a single region of interest, add the <code>start_position</code> and <code>end_position</code> arguments (units in megabases).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           chr_id <span class="op">=</span> <span class="fl">6</span>, 
           start_position <span class="op">=</span> <span class="fl">36.5</span>, 
           end_position <span class="op">=</span> <span class="fl">40.6</span>, 
           width_1 <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>
<span class="co">#&gt; Preparing plot data...</span>
<span class="co">#&gt; [1] "Task done, plot was stored in working directory."</span></code></pre></div>
<div class="figure">
<img src="roh_visual/Chr6_36.5-40.6Mb_roh.png" width="300" alt=""><p class="caption">Figure.25 ROH on Chr6:35.0-42.5 Mb</p>
</div>
</div>
<div id="how-can-we-present-annotated-genes-above-the-roh-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#how-can-we-present-annotated-genes-above-the-roh-plot" class="anchor"></a>16.3 How can we present annotated genes above the ROH plot?</h3>
<p>To add a panel of gene locations, add the <code>refgene</code> argument, pointing to a file containing genes annotated on the appropriate reference genome.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           chr_id <span class="op">=</span> <span class="fl">6</span>, 
           start_position <span class="op">=</span> <span class="fl">36.5</span>, 
           end_position <span class="op">=</span> <span class="fl">40.61</span>, 
           refgene <span class="op">=</span> <span class="st">"refgene/cow_UMD_UCSC.txt"</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 2 rows containing missing values (geom_vline).</span></code></pre></div>
<div class="figure">
<img src="roh_visual/Chr6_36.5-40.61Mb_roh.png" width="300" alt=""><p class="caption">Figure.26 Annotation of ROH on Chr6:36.5-40.5</p>
</div>
<p>If the ROH region has too many genes, plotting all the gene names will make the figure look too busy. We can choose the most important region(s) to plot using the <code>show_names</code> argument. In this example, we set two regions with the following argument: <code>show_name = c(37.28, 37.49, 38.25, 39.52)</code>. This will show the gene names in two regions: (37.28 - 37.49 Mb) and (38.25 - 39.52 Mb).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           chr_id <span class="op">=</span> <span class="fl">6</span>, 
           start_position <span class="op">=</span> <span class="fl">36.5</span>, 
           end_position <span class="op">=</span> <span class="fl">40.5</span>, 
           refgene <span class="op">=</span> <span class="st">"refgene/cow_UMD_UCSC.txt"</span>, 
           show_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">37.28</span>, <span class="fl">37.49</span>, <span class="fl">38.25</span>, <span class="fl">39.52</span><span class="op">)</span>, col_short <span class="op">=</span> <span class="st">"black"</span>, col_mid <span class="op">=</span> <span class="st">"green"</span>, col_long <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="roh_visual/Chr6_36.5-40.5Mb_roh.png" width="300" alt=""><p class="caption">Figure.27 Customised gene annotations on Chr6:36.5-40.5 Mb</p>
</div>
</div>
<div id="how-do-we-plot-the-full-length-of-roh-for-all-samples-that-had-a-roh-in-a-given-region" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-plot-the-full-length-of-roh-for-all-samples-that-had-a-roh-in-a-given-region" class="anchor"></a>16.4 How do we plot the full length of ROH for all samples that had a ROH in a given region?</h3>
<p>Above, We have plotted ROH in an interested region, but that method can only plot short ROHs which located in the given region, it will remove some longer ROHs which are partially overlapped to the given interval, that method can made a nice picture but will lost some samples, it still useful because we already know the sample size in that region. But if we want to present the whole length of ROH overlapped to our interested region (Such as a region of target gene), we can assign a vector to ‘target_region’ argument in ‘roh_visual’ function.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           target_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>,<span class="fl">38.25</span>,<span class="fl">39.52</span><span class="op">)</span>, <span class="co">#within the vector, the first integer indicate Chromosome. the rest are the start and end point, unit in megabases</span>
           col_short <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional</span>
           col_mid <span class="op">=</span> <span class="st">"gold"</span>, <span class="co">#optional</span>
           col_long <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional</span>
           height_1 <span class="op">=</span> <span class="fl">12</span>, <span class="co">#optional</span>
           width_1 <span class="op">=</span> <span class="fl">14</span><span class="op">)</span>
<span class="co">#&gt; Preparing plot data...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<div class="figure">
<img src="roh_visual/Chr6_38.25_39.52_roh.png" width="300" alt=""><p class="caption">Figure.28 ROHs with full length on target region Chr6:38.25-39.52 Mb</p>
</div>
</div>
<div id="how-do-we-plot-rohs-on-an-interested-sample" class="section level3">
<h3 class="hasAnchor">
<a href="#how-do-we-plot-rohs-on-an-interested-sample" class="anchor"></a>16.5 How do we plot ROHs on an interested sample?</h3>
<p>When we have some interested individuals, we can plot all the ROHs for these samples separately by setting the ‘individual_id’ arguments as following:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/roh_visual.html">roh_visual</a></span><span class="op">(</span>clean_roh <span class="op">=</span> <span class="st">"cnv_clean/cnvpart_roh.cnv"</span>, 
           individual_id <span class="op">=</span> <span class="st">"201094560008_R04C01"</span>, 
           height_1 <span class="op">=</span> <span class="fl">14</span>, <span class="co">#optional </span>
           width_1 <span class="op">=</span> <span class="fl">22</span>,  <span class="co">#optional</span>
           col_short <span class="op">=</span> <span class="st">"green"</span>, <span class="co">#optional</span>
           col_mid <span class="op">=</span> <span class="st">"yellow"</span>, <span class="co">#optional</span>
           col_long <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>
<span class="co">#&gt; Preparing plot data...</span>
<span class="co">#&gt; Task done, plot was stored in working directory.</span></code></pre></div>
<p><img src="HandyCNV_files/figure-html/unnamed-chunk-41-1.png" width="700"><img src="roh_visual/ROH_of_201094560008_R04C01.png" width="300" alt="Figure 29. ROHs on interested Sample"></p>
</div>
</div>
<div id="how-do-we-get-haplotype-for-roh-region" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-get-haplotype-for-roh-region" class="anchor"></a>17. How do we get haplotype for ROH region?</h2>
<p>ROH results will return us some long regions, but we may interested on some specific genes in the ROH region that experienced strong selection. Get the haplotypes of these genes in our experimental population may help us to find out some interesting patterns. we known that the ROH region generated by SNPs may be discontinuous due to insufficient marker density, beside, when infer ROH we considered the influence of genotyping quality by setting the minimal window size, accepting some missing value or maximum heterozygosity SNPs in the window, these conditions will exclude the short ROH regions in the final results. For instance, when we get the haplotype for a short region (such as a Gene), the frequency of homozygosity haplotypes may differing to the original ROH region it located.</p>
<p>The ‘get_haplotype’ function can help us to get the haplotype for all samples by given region. But now the application is limited, it only accept the phased genotype by Beagle (Version 5.1) with physical position, we may expend the compatibility later if necessary. The standard format of phased genotype and demo data used at here can be download from our github repository (<a href="https://github.com/JH-Zhou/HandyCNV/tree/master/vignettes/OARS_chr6" class="uri">https://github.com/JH-Zhou/HandyCNV/tree/master/vignettes/OARS_chr6</a>). Now, let’s go ahead and make haplotypes.</p>
<p>Extracting the haplotype will take three steps and combine three functions. The first step is to prepare the standard format for ‘HandyCNV’. We can use ‘prepare_geno’ function to prepare the phased genotype. We can also assign the ‘convert_letter = TRUE’ in ‘pre_phased’ function to convert the format of haploids from 0,1 to letters. See code as following. After running, it will return three list files contain the map, genotype and the columns index file.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genotype_by_number</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prep_phased.html">prep_phased</a></span><span class="op">(</span>phased_geno <span class="op">=</span> <span class="st">"OARS_chr6.vcf"</span><span class="op">)</span> <span class="co">#'OARS_chr6.vcf' is the demo data, need to download and save it in your working directory </span>
<span class="co">#&gt; Reading input file...</span>
<span class="co">#&gt; Spliting Sample columns...</span>
<span class="co">#&gt; Tansposing genotype...</span>

<span class="va">genotype_by_letter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prep_phased.html">prep_phased</a></span><span class="op">(</span>phased_geno <span class="op">=</span> <span class="st">"OARS_chr6.vcf"</span>, convert_letter <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co">#'OARS_chr6.vcf' is the demo data, need to download and save it in your working</span>
<span class="co">#&gt; Reading input file...</span>
<span class="co">#&gt; Coverting haplotype from 0, 1 to Ref and ALT...</span>
<span class="co">#&gt; Spliting Sample columns...</span>
<span class="co">#&gt; Tansposing genotype...</span></code></pre></div>
<p>Second step, we need to assign the interested region. Here we take the <em>NCAPG</em> gene as example, it located on BTA6:37,332,024-37,378,129 bp. The idea in this step is to find out the closest SNPs in this SNP chip.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/closer_snp.html">closer_snp</a></span><span class="op">(</span>phased_input <span class="op">=</span> <span class="va">genotype_by_letter</span><span class="op">$</span><span class="va">map</span>, <span class="co">#the map file in the list generated in 'pre_phased' function above</span>
                     chr <span class="op">=</span> <span class="fl">6</span>, start <span class="op">=</span> <span class="fl">37.332024</span>, end <span class="op">=</span> <span class="fl">37.378129</span><span class="op">)</span> <span class="co">#physical position unit in 'megabases'</span>
<span class="co">#&gt; # A tibble: 1 x 2</span>
<span class="co">#&gt;        min      max</span>
<span class="co">#&gt;      &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1 37340937 37377606</span></code></pre></div>
<p>Third step to get haplotype. It will return two lists contains the reocding haplotype and the index of recode type and the original haploid.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haplotype_number</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_haplotype.html">get_haplotype</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">genotype_by_number</span>, pos <span class="op">=</span> <span class="va">region</span><span class="op">)</span>
<span class="co">#&gt; [1] 2481</span>
<span class="co">#&gt; [1] 2490</span>
<span class="co">#&gt; There are 10 SNPs in this region...</span>
<span class="co">#&gt; Recoding Haploid...</span>
<span class="co">#&gt; Generating Haplotype...</span>

<span class="va">haplotype_letter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_haplotype.html">get_haplotype</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">genotype_by_letter</span>, pos <span class="op">=</span> <span class="va">region</span><span class="op">)</span>
<span class="co">#&gt; [1] 2481</span>
<span class="co">#&gt; [1] 2490</span>
<span class="co">#&gt; There are 10 SNPs in this region...</span>
<span class="co">#&gt; Recoding Haploid...</span>
<span class="co">#&gt; Generating Haplotype...</span></code></pre></div>
<p>Now, we can count the frequency of each haplotype in our population and to check what the haploids are in these haplotypes.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">haplotype_number</span><span class="op">$</span><span class="va">hap_type</span><span class="op">$</span><span class="va">Recode_Type</span>, dnn <span class="op">=</span> <span class="st">"Haplotype"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    Haplotype Freq</span>
<span class="co">#&gt; 1        1-1  353</span>
<span class="co">#&gt; 2        1-2   26</span>
<span class="co">#&gt; 3        1-4    5</span>
<span class="co">#&gt; 4        1-9    1</span>
<span class="co">#&gt; 5      10-11    1</span>
<span class="co">#&gt; 6        2-2    1</span>
<span class="co">#&gt; 7        3-3    6</span>
<span class="co">#&gt; 8        5-5    1</span>
<span class="co">#&gt; 9        6-6    1</span>
<span class="co">#&gt; 10       7-7    1</span>
<span class="co">#&gt; 11       8-8    1</span>
<span class="va">haplotype_number</span><span class="op">$</span><span class="va">hap_index</span>
<span class="co">#&gt; # A tibble: 11 x 3</span>
<span class="co">#&gt;    Haploid     Freq Re_code</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;int&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 0000000000   738       1</span>
<span class="co">#&gt;  2 1111111111    28       2</span>
<span class="co">#&gt;  3 000000N000    12       3</span>
<span class="co">#&gt;  4 1111101100     5       4</span>
<span class="co">#&gt;  5 00000N0000     2       5</span>
<span class="co">#&gt;  6 0N00000000     2       6</span>
<span class="co">#&gt;  7 0N0000N000     2       7</span>
<span class="co">#&gt;  8 N000000000     2       8</span>
<span class="co">#&gt;  9 0000001000     1       9</span>
<span class="co">#&gt; 10 0N0000000N     1      10</span>
<span class="co">#&gt; 11 0N0000100N     1      11</span>


<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">haplotype_letter</span><span class="op">$</span><span class="va">hap_type</span><span class="op">$</span><span class="va">Recode_Type</span>, dnn <span class="op">=</span> <span class="st">"Haplotype"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    Haplotype Freq</span>
<span class="co">#&gt; 1        1-1  353</span>
<span class="co">#&gt; 2        1-2   26</span>
<span class="co">#&gt; 3        1-4    5</span>
<span class="co">#&gt; 4        1-9    1</span>
<span class="co">#&gt; 5      10-11    1</span>
<span class="co">#&gt; 6        2-2    1</span>
<span class="co">#&gt; 7        3-3    6</span>
<span class="co">#&gt; 8        5-5    1</span>
<span class="co">#&gt; 9        6-6    1</span>
<span class="co">#&gt; 10       7-7    1</span>
<span class="co">#&gt; 11       8-8    1</span>
<span class="va">haplotype_letter</span><span class="op">$</span><span class="va">hap_index</span>
<span class="co">#&gt; # A tibble: 11 x 3</span>
<span class="co">#&gt;    Haploid     Freq Re_code</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;int&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ACAAGAGGAA   738       1</span>
<span class="co">#&gt;  2 GACGAGAAGG    28       2</span>
<span class="co">#&gt;  3 ACAAGANGAA    12       3</span>
<span class="co">#&gt;  4 GACGAAAAAA     5       4</span>
<span class="co">#&gt;  5 ACAAGNGGAA     2       5</span>
<span class="co">#&gt;  6 ANAAGAGGAA     2       6</span>
<span class="co">#&gt;  7 ANAAGANGAA     2       7</span>
<span class="co">#&gt;  8 NCAAGAGGAA     2       8</span>
<span class="co">#&gt;  9 ACAAGAAGAA     1       9</span>
<span class="co">#&gt; 10 ANAAGAAGAN     1      10</span>
<span class="co">#&gt; 11 ANAAGAGGAN     1      11</span></code></pre></div>
</div>
<div id="how-do-we-convert-coordinates-for-cnv-cnvr-roh-or-any-other-intervals" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-convert-coordinates-for-cnv-cnvr-roh-or-any-other-intervals" class="anchor"></a>18. How do we convert coordinates for CNV, CNVR, ROH, or any other intervals?</h2>
<p>The function <code>convert_coord</code> can used to convert coordinates for lists of intervals. Currently, the function is limited to inputs generated by the <code>convert_map</code> function, and can only convert the coordinates for intervals on the same type of SNP Chip. Converting coordinates may change the total length of the intervals, as of the positions and orders of the SNPs on the chromosome will potentially differ between various versions of reference genome; therefore, we should pay attention to the quality of the conversion, and use the results with caution. The function produces a summary table that summarises how many intervals were converted successfully, and reports on the differences in length between the converted and original intervals.</p>
<p>The <code>proxy_range</code> argument supports adjusting the genomic range to search and assign “proxy” locations for markers which are missing in the target map, in order to try to best reduce the loss of intervals after conversion. The principal is try to reduce the length of the intervals that were assigned to the proxy locations. For missing markers at the start positions, the search will begin at the missing position and proceed downsteam, selecting the closest suitable marker as the proxy position. For the end position, the search will begin from the missing end interval and proceed upsteam, again assigning the closest suitable marker as the proxy position. If no suitable markers can be found within <code>proxy_range</code> basepairs, the interval will be classed at “Missing”.</p>
<p>The optional argument <code>len_diff_threshold</code> is a quality control measure on the newly converted intervals that imposes a maximum difference of length between converted and original intervals. Only intervals with remapped positions within <code>len_diff_threshold</code> basepairs of the original position will pass the threshold.</p>
<p>Here, we convert the CNVPartition CNVR results from UMD3.1 coordinates to ARS1.2. The outputs are saved into the <code>convert_coord_penn_cnv</code> folder, and summary statistics about the conversion are displayed below.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/convert_coord.html">convert_coord</a></span><span class="op">(</span>input_def <span class="op">=</span> <span class="st">"cnv_clean/cleancnv.cnv"</span>, 
              def_tar_map <span class="op">=</span> <span class="st">"convert_map/def_tar_map.map"</span>,
              proxy_range <span class="op">=</span> <span class="fl">100000</span>,
              len_diff_threshold <span class="op">=</span> <span class="fl">100000</span>,
              folder <span class="op">=</span> <span class="st">"convert_coord_penn_cnv"</span><span class="op">)</span>
<span class="co">#&gt; [1] "'def_tar_map' file passed input check..."</span>
<span class="co">#&gt; [1] "The quality check of converion on Start Position as below: FALSE means not match between two version."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Match Mismatch </span>
<span class="co">#&gt;     4396      180 </span>
<span class="co">#&gt; [1] "The quality check of converion on End Position as below: FALSE means not match between two version."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Match Mismatch </span>
<span class="co">#&gt;     4322      254 </span>
<span class="co">#&gt; [1] "4104 Intervals passed quality check."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Fail Pass </span>
<span class="co">#&gt;  472 4104 </span>
<span class="co">#&gt;                                  Item    Length Number</span>
<span class="co">#&gt; 1                      Total Original 974963676   4576</span>
<span class="co">#&gt; 2                   Succeed Original  891042504   4104</span>
<span class="co">#&gt; 3                  Succeed Convertion 897762659      -</span>
<span class="co">#&gt; 4                   Failed convertion  77201017    472</span>
<span class="co">#&gt; 5 Difference Converted minus Original  -6720155      -</span></code></pre></div>
<p>We can also convert coordinates from the target map to the default map. For example, let’s convert the PennCNV ARS results to UMD 3.1. After converting the coordinates of CNVs or CNVRs, we could use the results which passed the quality check to make comparisons with the <code>compare_cnv</code> and <code>compare_cnvr</code> functions as section 8 and 9 described.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/convert_coord.html">convert_coord</a></span><span class="op">(</span>input_def <span class="op">=</span> <span class="st">"call_cnvr_standard/cnvr.txt"</span>, 
              proxy_range <span class="op">=</span> <span class="fl">100000</span>,
              len_diff_threshold <span class="op">=</span> <span class="fl">100000</span>,
              def_tar_map <span class="op">=</span> <span class="st">"convert_map/def_tar_map.map"</span>, 
              folder <span class="op">=</span> <span class="st">"convert_coord_penn_cnvR"</span><span class="op">)</span>
<span class="co">#&gt; [1] "'def_tar_map' file passed input check..."</span>
<span class="co">#&gt; [1] "The quality check of converion on Start Position as below: FALSE means not match between two version."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Match Mismatch </span>
<span class="co">#&gt;      871       10 </span>
<span class="co">#&gt; [1] "The quality check of converion on End Position as below: FALSE means not match between two version."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Match Mismatch </span>
<span class="co">#&gt;      870       11 </span>
<span class="co">#&gt; [1] "835 Intervals passed quality check."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Fail Pass </span>
<span class="co">#&gt;   46  835 </span>
<span class="co">#&gt;                                  Item    Length Number</span>
<span class="co">#&gt; 1                      Total Original 286942723    881</span>
<span class="co">#&gt; 2                   Succeed Original  270103909    835</span>
<span class="co">#&gt; 3                  Succeed Convertion 271977583      -</span>
<span class="co">#&gt; 4                   Failed convertion  14965140     46</span>
<span class="co">#&gt; 5 Difference Converted minus Original  -1873674      -</span></code></pre></div>
</div>
<div id="how-do-we-make-a-fast-snp-density-plot-from-plink-map-file" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-we-make-a-fast-snp-density-plot-from-plink-map-file" class="anchor"></a>19. How do we make a fast SNP density plot from Plink map file?</h2>
<p>Plotting SNP density may help us learn about the SNP distribution on chromosomes. The parameter of SNP density may affect ROH detection in some tools. ‘plot_snp_density’ only working with map file that has four columns in the order as follows: Chromosome, SNP ID, Position in Morgans, Physical Position. No header required.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_snp_density.html">plot_snp_density</a></span><span class="op">(</span>map <span class="op">=</span> <span class="st">"convert_map/target_plink.map"</span>, 
                 max_chr <span class="op">=</span> <span class="fl">24</span>, <span class="co">#optional</span>
                 top_density <span class="op">=</span> <span class="fl">60</span>, <span class="co">#optional </span>
                 low_density <span class="op">=</span> <span class="fl">20</span>, <span class="co">#optional</span>
                 color_top <span class="op">=</span> <span class="st">"red"</span>, <span class="co">#optional</span>
                 color_low <span class="op">=</span> <span class="st">"blue"</span>, <span class="co">#optional</span>
                 color_mid <span class="op">=</span> <span class="st">"black"</span>, <span class="co">#optional</span>
                 legend_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="co">#optional </span>
                 x_label <span class="op">=</span> <span class="st">"Physical position\n&lt;U+7269&gt;&lt;U+7406&gt;&lt;U+4F4D&gt;&lt;U+7F6E&gt;"</span>NA<span class="co">#optional</span>NA<span class="op">=</span> <span class="st">"SNPs/Mb\n&lt;U+6BCF&gt;1Mb&lt;U+533A&gt;&lt;U+95F4&gt;&lt;U+7684&gt;SNP&lt;U+6570&gt;"</span>NA<span class="co">#optional</span>NA<span class="op">=</span> <span class="fl">5</span><span class="op">)</span> 
<span class="co">#&gt; The total number of SNPs on selected chromosomes are 119249</span>
<span class="co">#&gt; The number of regions on different SNP density groups as following:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  &lt;=20  &gt;=60 20-60 </span>
<span class="co">#&gt;    39   397  1849 </span>
<span class="co">#&gt; The Mean of SNPs density is  52  SNPs/Mb</span>
<span class="co">#&gt; The Mean + SD of SNP density is  73  SNPs/Mb</span>
<span class="co">#&gt; The Mean - SD of one SD is  31  SNPs/Mb</span>
<span class="co">#save the plot by 'ggsave'</span>
<span class="co">#ggsave(filename = "snp_density.png", width = 26, height = 18, units = "cm", dpi = 350)</span></code></pre></div>
<div class="figure">
<img src="SNP_density.png" width="300" alt=""><p class="caption">Figure 30. SNP density distribution</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jinghang Zhou, Liyuan Liu, Thomas Lopdell, Dorian Garrick, Yuangang Shi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
